{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dask import diagnostics, widgets\n",
    "import pandas as pd\n",
    "from pandas_plink import read_plink\n",
    "from pandas_plink import get_data_folder\n",
    "import numpy as np\n",
    "import os\n",
    "from os.path import join\n",
    "import gzip\n",
    "from io import StringIO\n",
    "import glob\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cis-eQTL for one chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_window = 500000\n",
    "chrom_num = 22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nbim_combined = []\\nfam_combined = []\\nbed_combined = []\\n\\nfor i in range(1, 23):  \\n\\n    file_base = f\"{file_path}/1000G.EUR.{i}\"  \\n\\n    # Read the BIM, FAM, BED files\\n    bim, fam, bed = read_plink(file_base)\\n    \\n    # Append to combined lists\\n    bim_combined.append(bim)\\n    fam_combined.append(fam)\\n    bed_combined.append(bed)\\n\\n\\n'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pandas_plink import read_plink\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "\n",
    "# Base path to the files\n",
    "file_path = \"./LDREF\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 24.64it/s]\n"
     ]
    }
   ],
   "source": [
    "file_base = f\"{file_path}/1000G.EUR.{chrom_num}\"  \n",
    "bim, fam, bed = read_plink(file_base)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp_df = pd.read_csv('GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz', sep='\\t', compression='gzip', header=0)\n",
    "\n",
    "# Find intersection with gene expression data\n",
    "gene_exp_samples = gene_exp_df.columns[4:]\n",
    "matched_samples = set(gene_exp_samples).intersection(fam['iid'])\n",
    "\n",
    "# Filter gene expression data\n",
    "filtered_gene_exp = gene_exp_df[['TargetID', 'Gene_Symbol', 'Chr', 'Coord'] + list(matched_samples)]\n",
    "filtered_gene_exp.loc[:,'Gene_Symbol'] = filtered_gene_exp['Gene_Symbol'].apply(lambda x: x.split('.')[0])\n",
    "gene_annot = pd.read_csv('gene_annot.txt.gz', sep='\\t', compression='gzip', header=0)\n",
    "\n",
    "combined_filtered_gene_exp = pd.merge(gene_annot, filtered_gene_exp, left_on='SYM', right_on='Gene_Symbol').drop(columns=['TargetID', 'Gene_Symbol', 'Chr'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp_chrom = combined_filtered_gene_exp[combined_filtered_gene_exp['CHR'] == chrom_num]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_indices = fam[fam['iid'].isin(gene_exp_chrom.columns)]['i'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "344"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sample_indices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a0</th>\n",
       "      <th>a1</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>rs9617528</td>\n",
       "      <td>0.059576</td>\n",
       "      <td>16061016</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>rs4911642</td>\n",
       "      <td>2.127458</td>\n",
       "      <td>16504399</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>rs7287144</td>\n",
       "      <td>2.252105</td>\n",
       "      <td>16886873</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>rs5748662</td>\n",
       "      <td>2.255191</td>\n",
       "      <td>16892858</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>rs5994034</td>\n",
       "      <td>2.255836</td>\n",
       "      <td>16894090</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom        snp        cm       pos a0 a1  i\n",
       "0    22  rs9617528  0.059576  16061016  C  T  0\n",
       "1    22  rs4911642  2.127458  16504399  C  T  1\n",
       "2    22  rs7287144  2.252105  16886873  G  A  2\n",
       "3    22  rs5748662  2.255191  16892858  A  G  3\n",
       "4    22  rs5994034  2.255836  16894090  T  C  4"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim_chrom = bim[bim['chrom'] == str(chrom_num)] \n",
    "bim_chrom.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fid</th>\n",
       "      <th>iid</th>\n",
       "      <th>father</th>\n",
       "      <th>mother</th>\n",
       "      <th>gender</th>\n",
       "      <th>trait</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00096</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00097</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00099</td>\n",
       "      <td>HG00099</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00101</td>\n",
       "      <td>HG00101</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00102</td>\n",
       "      <td>HG00102</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fid      iid father mother gender trait  i\n",
       "0  HG00096  HG00096      0      0      0    -9  0\n",
       "1  HG00097  HG00097      0      0      0    -9  1\n",
       "2  HG00099  HG00099      0      0      0    -9  2\n",
       "3  HG00101  HG00101      0      0      0    -9  3\n",
       "4  HG00102  HG00102      0      0      0    -9  4"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fam_chrom = fam[fam['iid'].isin(gene_exp_chrom.columns)].drop_duplicates()\n",
    "fam_chrom.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>STOP</th>\n",
       "      <th>SYM</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>Coord</th>\n",
       "      <th>HG00320</th>\n",
       "      <th>HG00109</th>\n",
       "      <th>NA11992</th>\n",
       "      <th>...</th>\n",
       "      <th>HG00232</th>\n",
       "      <th>NA12827</th>\n",
       "      <th>HG00383</th>\n",
       "      <th>NA20774</th>\n",
       "      <th>NA20805</th>\n",
       "      <th>HG00321</th>\n",
       "      <th>HG00284</th>\n",
       "      <th>HG00111</th>\n",
       "      <th>NA12751</th>\n",
       "      <th>HG00252</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14076</th>\n",
       "      <td>TPTEP1</td>\n",
       "      <td>22</td>\n",
       "      <td>16601886</td>\n",
       "      <td>16698742</td>\n",
       "      <td>ENSG00000100181</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>17082777</td>\n",
       "      <td>0.185925</td>\n",
       "      <td>0.203435</td>\n",
       "      <td>0.054817</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027105</td>\n",
       "      <td>0.032731</td>\n",
       "      <td>0.154169</td>\n",
       "      <td>0.178876</td>\n",
       "      <td>0.021584</td>\n",
       "      <td>0.138682</td>\n",
       "      <td>0.109612</td>\n",
       "      <td>0.152036</td>\n",
       "      <td>0.189095</td>\n",
       "      <td>0.093371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14077</th>\n",
       "      <td>CECR7</td>\n",
       "      <td>22</td>\n",
       "      <td>17036569</td>\n",
       "      <td>17060825</td>\n",
       "      <td>ENSG00000237438</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>17517460</td>\n",
       "      <td>1.123570</td>\n",
       "      <td>0.738311</td>\n",
       "      <td>1.532103</td>\n",
       "      <td>...</td>\n",
       "      <td>0.640684</td>\n",
       "      <td>3.996098</td>\n",
       "      <td>0.438416</td>\n",
       "      <td>0.501395</td>\n",
       "      <td>1.878676</td>\n",
       "      <td>0.470452</td>\n",
       "      <td>0.932302</td>\n",
       "      <td>1.216698</td>\n",
       "      <td>-0.138827</td>\n",
       "      <td>1.235000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14078</th>\n",
       "      <td>IL17RA</td>\n",
       "      <td>22</td>\n",
       "      <td>17084953</td>\n",
       "      <td>17115693</td>\n",
       "      <td>ENSG00000177663</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>17565844</td>\n",
       "      <td>2.580380</td>\n",
       "      <td>3.773908</td>\n",
       "      <td>2.990411</td>\n",
       "      <td>...</td>\n",
       "      <td>2.532821</td>\n",
       "      <td>3.130056</td>\n",
       "      <td>6.516216</td>\n",
       "      <td>6.954348</td>\n",
       "      <td>2.500827</td>\n",
       "      <td>4.069094</td>\n",
       "      <td>2.773805</td>\n",
       "      <td>3.125675</td>\n",
       "      <td>2.483425</td>\n",
       "      <td>3.326086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14079</th>\n",
       "      <td>CECR6</td>\n",
       "      <td>22</td>\n",
       "      <td>17116298</td>\n",
       "      <td>17121367</td>\n",
       "      <td>ENSG00000183307</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>17602257</td>\n",
       "      <td>0.010100</td>\n",
       "      <td>0.073524</td>\n",
       "      <td>0.068237</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053998</td>\n",
       "      <td>0.027648</td>\n",
       "      <td>0.045082</td>\n",
       "      <td>0.075619</td>\n",
       "      <td>0.041435</td>\n",
       "      <td>0.015151</td>\n",
       "      <td>0.048417</td>\n",
       "      <td>0.046828</td>\n",
       "      <td>0.077449</td>\n",
       "      <td>0.017031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14080</th>\n",
       "      <td>CECR5</td>\n",
       "      <td>22</td>\n",
       "      <td>17137510</td>\n",
       "      <td>17165287</td>\n",
       "      <td>ENSG00000069998</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>17646177</td>\n",
       "      <td>27.267517</td>\n",
       "      <td>27.230939</td>\n",
       "      <td>38.081339</td>\n",
       "      <td>...</td>\n",
       "      <td>31.605359</td>\n",
       "      <td>25.771682</td>\n",
       "      <td>27.386406</td>\n",
       "      <td>22.612313</td>\n",
       "      <td>32.924733</td>\n",
       "      <td>31.770295</td>\n",
       "      <td>27.900993</td>\n",
       "      <td>30.982620</td>\n",
       "      <td>18.655526</td>\n",
       "      <td>24.509548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14450</th>\n",
       "      <td>CHKB-AS1</td>\n",
       "      <td>22</td>\n",
       "      <td>50583025</td>\n",
       "      <td>50583877</td>\n",
       "      <td>ENSG00000205559</td>\n",
       "      <td>antisense</td>\n",
       "      <td>51021455</td>\n",
       "      <td>0.339837</td>\n",
       "      <td>0.493989</td>\n",
       "      <td>0.617282</td>\n",
       "      <td>...</td>\n",
       "      <td>0.296301</td>\n",
       "      <td>0.386448</td>\n",
       "      <td>0.399380</td>\n",
       "      <td>0.259015</td>\n",
       "      <td>0.424392</td>\n",
       "      <td>0.950139</td>\n",
       "      <td>0.705295</td>\n",
       "      <td>0.217012</td>\n",
       "      <td>0.789061</td>\n",
       "      <td>0.619564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14451</th>\n",
       "      <td>ARSA</td>\n",
       "      <td>22</td>\n",
       "      <td>50622753</td>\n",
       "      <td>50628179</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>51066607</td>\n",
       "      <td>5.100634</td>\n",
       "      <td>4.850561</td>\n",
       "      <td>4.489545</td>\n",
       "      <td>...</td>\n",
       "      <td>4.030331</td>\n",
       "      <td>6.845832</td>\n",
       "      <td>5.100789</td>\n",
       "      <td>5.266580</td>\n",
       "      <td>2.160187</td>\n",
       "      <td>7.427782</td>\n",
       "      <td>3.887837</td>\n",
       "      <td>4.493735</td>\n",
       "      <td>6.784537</td>\n",
       "      <td>5.521356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14452</th>\n",
       "      <td>SHANK3</td>\n",
       "      <td>22</td>\n",
       "      <td>50674414</td>\n",
       "      <td>50733298</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>51113070</td>\n",
       "      <td>0.016057</td>\n",
       "      <td>0.010864</td>\n",
       "      <td>0.014591</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003729</td>\n",
       "      <td>0.011049</td>\n",
       "      <td>0.010625</td>\n",
       "      <td>0.017023</td>\n",
       "      <td>0.014645</td>\n",
       "      <td>-0.003839</td>\n",
       "      <td>0.014828</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>0.002815</td>\n",
       "      <td>0.001015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14453</th>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>22</td>\n",
       "      <td>50756947</td>\n",
       "      <td>50801309</td>\n",
       "      <td>ENSG00000184319</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>51195376</td>\n",
       "      <td>12.595851</td>\n",
       "      <td>13.827377</td>\n",
       "      <td>15.841559</td>\n",
       "      <td>...</td>\n",
       "      <td>13.525351</td>\n",
       "      <td>28.951106</td>\n",
       "      <td>12.462425</td>\n",
       "      <td>18.389374</td>\n",
       "      <td>15.800450</td>\n",
       "      <td>16.175383</td>\n",
       "      <td>16.228767</td>\n",
       "      <td>10.098311</td>\n",
       "      <td>11.893010</td>\n",
       "      <td>20.328535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14454</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>22</td>\n",
       "      <td>50767500</td>\n",
       "      <td>50783663</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>51222091</td>\n",
       "      <td>11.421610</td>\n",
       "      <td>11.451661</td>\n",
       "      <td>11.821741</td>\n",
       "      <td>...</td>\n",
       "      <td>9.628176</td>\n",
       "      <td>8.316205</td>\n",
       "      <td>10.135038</td>\n",
       "      <td>7.418261</td>\n",
       "      <td>15.428374</td>\n",
       "      <td>12.349890</td>\n",
       "      <td>10.988388</td>\n",
       "      <td>9.391654</td>\n",
       "      <td>11.281292</td>\n",
       "      <td>12.145755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>379 rows × 351 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID  CHR     START      STOP              SYM  \\\n",
       "14076     TPTEP1   22  16601886  16698742  ENSG00000100181   \n",
       "14077      CECR7   22  17036569  17060825  ENSG00000237438   \n",
       "14078     IL17RA   22  17084953  17115693  ENSG00000177663   \n",
       "14079      CECR6   22  17116298  17121367  ENSG00000183307   \n",
       "14080      CECR5   22  17137510  17165287  ENSG00000069998   \n",
       "...          ...  ...       ...       ...              ...   \n",
       "14450   CHKB-AS1   22  50583025  50583877  ENSG00000205559   \n",
       "14451       ARSA   22  50622753  50628179  ENSG00000100299   \n",
       "14452     SHANK3   22  50674414  50733298  ENSG00000251322   \n",
       "14453  RPL23AP82   22  50756947  50801309  ENSG00000184319   \n",
       "14454     RABL2B   22  50767500  50783663  ENSG00000079974   \n",
       "\n",
       "                                     TYPE     Coord    HG00320    HG00109  \\\n",
       "14076                             lincRNA  17082777   0.185925   0.203435   \n",
       "14077                             lincRNA  17517460   1.123570   0.738311   \n",
       "14078                      protein_coding  17565844   2.580380   3.773908   \n",
       "14079                      protein_coding  17602257   0.010100   0.073524   \n",
       "14080                      protein_coding  17646177  27.267517  27.230939   \n",
       "...                                   ...       ...        ...        ...   \n",
       "14450                           antisense  51021455   0.339837   0.493989   \n",
       "14451                      protein_coding  51066607   5.100634   4.850561   \n",
       "14452                      protein_coding  51113070   0.016057   0.010864   \n",
       "14453  transcribed_unprocessed_pseudogene  51195376  12.595851  13.827377   \n",
       "14454                      protein_coding  51222091  11.421610  11.451661   \n",
       "\n",
       "         NA11992  ...    HG00232    NA12827    HG00383    NA20774    NA20805  \\\n",
       "14076   0.054817  ...   0.027105   0.032731   0.154169   0.178876   0.021584   \n",
       "14077   1.532103  ...   0.640684   3.996098   0.438416   0.501395   1.878676   \n",
       "14078   2.990411  ...   2.532821   3.130056   6.516216   6.954348   2.500827   \n",
       "14079   0.068237  ...   0.053998   0.027648   0.045082   0.075619   0.041435   \n",
       "14080  38.081339  ...  31.605359  25.771682  27.386406  22.612313  32.924733   \n",
       "...          ...  ...        ...        ...        ...        ...        ...   \n",
       "14450   0.617282  ...   0.296301   0.386448   0.399380   0.259015   0.424392   \n",
       "14451   4.489545  ...   4.030331   6.845832   5.100789   5.266580   2.160187   \n",
       "14452   0.014591  ...  -0.003729   0.011049   0.010625   0.017023   0.014645   \n",
       "14453  15.841559  ...  13.525351  28.951106  12.462425  18.389374  15.800450   \n",
       "14454  11.821741  ...   9.628176   8.316205  10.135038   7.418261  15.428374   \n",
       "\n",
       "         HG00321    HG00284    HG00111    NA12751    HG00252  \n",
       "14076   0.138682   0.109612   0.152036   0.189095   0.093371  \n",
       "14077   0.470452   0.932302   1.216698  -0.138827   1.235000  \n",
       "14078   4.069094   2.773805   3.125675   2.483425   3.326086  \n",
       "14079   0.015151   0.048417   0.046828   0.077449   0.017031  \n",
       "14080  31.770295  27.900993  30.982620  18.655526  24.509548  \n",
       "...          ...        ...        ...        ...        ...  \n",
       "14450   0.950139   0.705295   0.217012   0.789061   0.619564  \n",
       "14451   7.427782   3.887837   4.493735   6.784537   5.521356  \n",
       "14452  -0.003839   0.014828   0.006472   0.002815   0.001015  \n",
       "14453  16.175383  16.228767  10.098311  11.893010  20.328535  \n",
       "14454  12.349890  10.988388   9.391654  11.281292  12.145755  \n",
       "\n",
       "[379 rows x 351 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_exp_chrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n",
      "c:\\Users\\linyo\\OneDrive\\Desktop\\UCSD\\CS_DS\\DSC 180\\myenv\\Lib\\site-packages\\dask\\core.py:133: RuntimeWarning: invalid value encountered in divide\n",
      "  return func(*(_execute_task(a, cache) for a in args))\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "\n",
    "# filtered_genotypes_df = pd.DataFrame(filtered_genotypes.compute() , index=sample_indices, columns=bim['snp'])\n",
    "\n",
    "results = []\n",
    "\n",
    "for i in range(len(gene_exp_chrom)):\n",
    "    gene = gene_exp_chrom.iloc[i]\n",
    "    \n",
    "    idx, start, stop = gene['ID'], gene['START'] - cis_window, gene['STOP'] + cis_window\n",
    "    expression = gene.iloc[7:].values.astype(float) # (num_samples, )\n",
    "\n",
    "    filtered_snps = bim_chrom[(bim_chrom['pos'] >= start) & (bim_chrom['pos'] <= stop)]\n",
    "    \n",
    "    snp_indices = filtered_snps['snp'].index.to_list()\n",
    "    filtered_genotypes = bed[snp_indices, :][:, sample_indices] # (num_snp, num_samples)\n",
    "    \n",
    "    snp_genotype_mean = np.mean(filtered_genotypes, axis=1, keepdims=True)  # (num_snp, 1)\n",
    "    expression_mean = np.mean(expression)  # scalar\n",
    "\n",
    "    centered_genotypes = filtered_genotypes - snp_genotype_mean\n",
    "    centered_expression = expression - expression_mean\n",
    "\n",
    "    numerator = np.dot(centered_genotypes, centered_expression)  # (num_snp,)\n",
    "    denominator = np.sum(centered_genotypes ** 2, axis=1)  # (num_snp,)\n",
    "\n",
    "    \n",
    "    slope = numerator / denominator  # (num_snp,)\n",
    "    intercept = expression_mean - slope * snp_genotype_mean[:, 0]  # (num_snp,)\n",
    "    \n",
    "    fitted_values = intercept[:, None] + slope[:, None] * filtered_genotypes # (num_snps, num_samples)\n",
    "    ss_total = np.sum((expression - expression_mean) ** 2) # scalar\n",
    "    ss_residual = np.sum((expression[None, :]  - fitted_values) ** 2, axis=1) # (num_snps, )\n",
    "    r_squared = 1 - (ss_residual / ss_total) # (num_snps, )\n",
    "    \n",
    "    dof = len(expression) - 2  \n",
    "    std_err = np.sqrt((ss_residual / dof) / denominator) # (num_snps, )\n",
    "    t_values = slope / std_err # (num_snps, )\n",
    "    p_values = 2 * stats.t.sf(np.abs(t_values), dof) # (num_snps, )\n",
    "    \n",
    "    batch_df = pd.DataFrame({\n",
    "        'chrom': chrom_num,              \n",
    "        'gene_id': idx,                    \n",
    "        'snp': filtered_snps['snp'].values,\n",
    "        'pos': filtered_snps['pos'].values,\n",
    "        'ref': filtered_snps['a0'].values,  \n",
    "        'alt': filtered_snps['a1'].values, \n",
    "        'beta': slope,  \n",
    "        'se': std_err,                             \n",
    "        'r_squared': r_squared,           \n",
    "        'p_value': p_values,                        \n",
    "    })\n",
    "\n",
    "    results.append(batch_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       chrom gene_id         snp       pos ref alt      beta        se  \\\n",
      "32428     22    GNAZ   rs6003561  23558472   T   C -1.099425  0.123069   \n",
      "43119     22    UPB1   rs9624395  24460593   C   A -1.929328  0.353915   \n",
      "43163     22    UPB1  rs11090310  24569468   T   C -1.929328  0.353915   \n",
      "43139     22    UPB1   rs9620358  24518086   G   A -1.929328  0.353915   \n",
      "43112     22    UPB1  rs17004823  24447303   A   G -1.929328  0.353915   \n",
      "\n",
      "       r_squared       p_value  \n",
      "32428   0.189200  2.601136e-17  \n",
      "43119   0.079947  9.588345e-08  \n",
      "43163   0.079947  9.588345e-08  \n",
      "43139   0.079947  9.588345e-08  \n",
      "43112   0.079947  9.588345e-08  \n"
     ]
    }
   ],
   "source": [
    "results_df = pd.concat(results, ignore_index=True)\n",
    "results_df = results_df.dropna(subset=['p_value'])\n",
    "print(results_df.sort_values('p_value').head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>snp</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>TPTEP1</td>\n",
       "      <td>rs4911642</td>\n",
       "      <td>16504399</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.004224</td>\n",
       "      <td>0.006564</td>\n",
       "      <td>0.001209</td>\n",
       "      <td>0.520345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>TPTEP1</td>\n",
       "      <td>rs7287144</td>\n",
       "      <td>16886873</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001396</td>\n",
       "      <td>0.005489</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.799451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>TPTEP1</td>\n",
       "      <td>rs5748662</td>\n",
       "      <td>16892858</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.003162</td>\n",
       "      <td>0.005299</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.551101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>TPTEP1</td>\n",
       "      <td>rs5994034</td>\n",
       "      <td>16894090</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.003893</td>\n",
       "      <td>0.007754</td>\n",
       "      <td>0.000737</td>\n",
       "      <td>0.615945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>TPTEP1</td>\n",
       "      <td>rs4010554</td>\n",
       "      <td>16894264</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000676</td>\n",
       "      <td>0.005365</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.899784</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom gene_id        snp       pos ref alt      beta        se  r_squared  \\\n",
       "0     22  TPTEP1  rs4911642  16504399   C   T -0.004224  0.006564   0.001209   \n",
       "1     22  TPTEP1  rs7287144  16886873   G   A  0.001396  0.005489   0.000189   \n",
       "2     22  TPTEP1  rs5748662  16892858   A   G  0.003162  0.005299   0.001040   \n",
       "3     22  TPTEP1  rs5994034  16894090   T   C  0.003893  0.007754   0.000737   \n",
       "4     22  TPTEP1  rs4010554  16894264   A   C  0.000676  0.005365   0.000046   \n",
       "\n",
       "    p_value  \n",
       "0  0.520345  \n",
       "1  0.799451  \n",
       "2  0.551101  \n",
       "3  0.615945  \n",
       "4  0.899784  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results saved to 'cis-eQTL_results.txt'\n"
     ]
    }
   ],
   "source": [
    "sorted_results = results_df.sort_values('pos')\n",
    "sorted_results.to_csv('cis-eQTL_results.txt', sep='\\t', index=False)\n",
    "\n",
    "print(\"Results saved to 'cis-eQTL_results.txt'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pruning and Thresholding (P + T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(Index(['chrom', 'gene_id', 'snp', 'pos', 'ref', 'alt', 'beta', 'se',\n",
       "        'r_squared', 'p_value'],\n",
       "       dtype='object'),\n",
       " Index(['chrom', 'snp', 'cm', 'pos', 'a0', 'a1', 'i'], dtype='object'))"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_results.columns, bim_chrom.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_chrom = bim_chrom.rename(columns={\"a0\": \"ref\", \"a1\": \"alt\"})\n",
    "bim_chrom['chrom'] = bim_chrom['chrom'].astype(int)\n",
    "\n",
    "clump_folder = \"./clumps\" \n",
    "valid_snp_folder = \"./valid-snp\" \n",
    "\n",
    "data_folders = [\"./prs_intermediate/clumps\", \"./prs_intermediate/valid-snp\", \"./prs_intermediate/indivs\", \"./prs_intermediate/train\", \"./results/prs_model_building/valid\", \"./results/prs_model_building/test\", \"./prs_intermediate/snp-pval\"]\n",
    "\n",
    "for folder in data_folders:\n",
    "    os.makedirs(folder, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clump(p_val, summary_stats = \"cis-eqtl_results.txt\", clump_kb = 250, clump_r2 = 0.1):\n",
    "    data = pd.merge(bim_chrom, sorted_results, on=[\"chrom\", \"snp\", \"pos\", \"ref\", \"alt\"])\n",
    "    data = data[data['p_value'] < 1]\n",
    "    \n",
    "    plink_command = (\n",
    "        f\".\\plink --bfile ./LDREF/1000G.EUR.{chrom_num} --clump-p1 {p_val} --clump-r2 {clump_r2} \"\n",
    "        f\"--clump-kb {clump_kb} --clump {summary_stats} \"\n",
    "        f\"--clump-snp-field snp --clump-field p_value --allow-no-sex --out {f'./prs_intermediate/clumps/EUR__{p_val}'}\"\n",
    "    )\n",
    "\n",
    "    \n",
    "    try:\n",
    "        result = subprocess.run(plink_command, shell=True, check=True, capture_output=True, text=True)\n",
    "        print(\"PLINK output:\", result.stdout)  \n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Error occurred:\", e.stderr)  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./clumps/EUR__0.05.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --clump cis-eqtl_results.txt\n",
      "  --clump-field p_value\n",
      "  --clump-kb 250\n",
      "  --clump-p1 0.05\n",
      "  --clump-r2 0.1\n",
      "  --clump-snp-field snp\n",
      "  --out ./clumps/EUR__0.05\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./clumps/EUR__0.05.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 489 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "17489 variants and 489 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--clump: 1118 clumps formed from 5922 top variants.\n",
      "Results written to ./clumps/EUR__0.05.clumped .\n",
      "\n"
     ]
    }
   ],
   "source": [
    "clump(0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def valid_snp_idx(p_val):\n",
    "    command = f\"\"\"\n",
    "        Get-Content ./prs_intermediate/clumps/EUR__{p_val}.clumped |\n",
    "        Select-Object -Skip 1 |\n",
    "        ForEach-Object {{ ($_ -split '\\\\s+')[3] }} |\n",
    "        Set-Content ./prs_intermediate/valid-snp/EUR__{p_val}.valid.snp\n",
    "    \"\"\"\n",
    "    \n",
    "    try:\n",
    "        # Run the command in PowerShell\n",
    "        subprocess.run([\"powershell\", \"-Command\", command], check=True)\n",
    "        print(\"Index SNP IDs extracted.\")\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Error occurred:\", e.stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index SNP IDs extracted.\n"
     ]
    }
   ],
   "source": [
    "valid_snp_idx(0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fid</th>\n",
       "      <th>iid</th>\n",
       "      <th>father</th>\n",
       "      <th>mother</th>\n",
       "      <th>gender</th>\n",
       "      <th>trait</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00096</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00097</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00099</td>\n",
       "      <td>HG00099</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00101</td>\n",
       "      <td>HG00101</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00102</td>\n",
       "      <td>HG00102</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>NA20814</td>\n",
       "      <td>NA20814</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>NA20815</td>\n",
       "      <td>NA20815</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>NA20819</td>\n",
       "      <td>NA20819</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>NA20826</td>\n",
       "      <td>NA20826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>NA20828</td>\n",
       "      <td>NA20828</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "      <td>487</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>344 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         fid      iid father mother gender trait    i\n",
       "0    HG00096  HG00096      0      0      0    -9    0\n",
       "1    HG00097  HG00097      0      0      0    -9    1\n",
       "2    HG00099  HG00099      0      0      0    -9    2\n",
       "3    HG00101  HG00101      0      0      0    -9    3\n",
       "4    HG00102  HG00102      0      0      0    -9    4\n",
       "..       ...      ...    ...    ...    ...   ...  ...\n",
       "479  NA20814  NA20814      0      0      0    -9  479\n",
       "480  NA20815  NA20815      0      0      0    -9  480\n",
       "482  NA20819  NA20819      0      0      0    -9  482\n",
       "485  NA20826  NA20826      0      0      0    -9  485\n",
       "487  NA20828  NA20828      0      0      0    -9  487\n",
       "\n",
       "[344 rows x 7 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fam_filtered = fam[fam['fid'].isin(matched_samples)]\n",
    "fam_filtered.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_val_test_ids():\n",
    "    fam_data = fam.drop_duplicates().sample(frac=1, random_state=42).reset_index(drop=True)\n",
    "\n",
    "    # Split into 80%, 10%, 10%\n",
    "    n_total = len(fam_data)\n",
    "    train_set = fam_data.iloc[:int(0.8 * n_total)][[\"fid\", \"iid\"]]\n",
    "    validation_set = fam_data.iloc[int(0.8 * n_total):int(0.9 * n_total)][[\"fid\", \"iid\"]]\n",
    "    test_set = fam_data.iloc[int(0.9 * n_total):][[\"fid\", \"iid\"]]\n",
    "    \n",
    "\n",
    "    # Save each subset\n",
    "    train_set.to_csv(\"./prs_intermediate/indivs/80_percent_ids.txt\", sep=\"\\t\", index=False, header=True)\n",
    "    validation_set.to_csv(\"./prs_intermediate/indivs/validation_set_ids.txt\", sep=\"\\t\", index=False, header=True)\n",
    "    test_set.to_csv(\"./prs_intermediate/indivs/test_set_ids.txt\", sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_exp_txt(gene_name, mode):\n",
    "    gene_data = gene_exp_chrom.set_index('ID').loc[gene_name][7:]\n",
    "\n",
    "    gene_df = pd.DataFrame({\n",
    "        \"fid\": [x for x in gene_data.index],  \n",
    "        \"iid\": [x for x in gene_data.index],  \n",
    "        \"Expression\": gene_data.values\n",
    "    })\n",
    "    \n",
    "    filt_ids = pd.DataFrame()\n",
    "    \n",
    "    if mode == 'train':\n",
    "        filt_ids = pd.read_csv(\"./prs_intermediate/indivs/80_percent_ids.txt\", sep='\\s+', header=None, names=[\"iid\"])\n",
    "    elif mode == 'valid':\n",
    "        filt_ids = pd.read_csv(\"./prs_intermediate/indivs/validation_set_ids.txt\", sep='\\s+', header=None, names=[\"iid\"])\n",
    "    else:\n",
    "        filt_ids = pd.read_csv(\"./prs_intermediate/indivs/test_set_ids.txt\", sep='\\s+', header=None, names=[\"iid\"])\n",
    "        \n",
    "    filtered_data = gene_df[gene_df[\"iid\"].isin(filt_ids[\"iid\"])]\n",
    "    \n",
    "    filtered_data.to_csv(f\"gene_exp_{gene_name}_{mode}.txt\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name = 'CECR5'\n",
    "train_val_test_ids()\n",
    "gene_exp_txt(gene_name=gene_name, mode='train')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cis_range(gene_name, window=500000):\n",
    "    gene_info = gene_exp_chrom.set_index('ID').loc[gene_name]\n",
    "    start, stop = gene_info['START'] - window, gene_info['STOP'] + window\n",
    "    return start, stop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_gene_cis_eqtl(gene_name, mode='train'):\n",
    "    \n",
    "    # Subset train\n",
    "    plink_command = (\n",
    "        f\".\\plink --bfile ./LDREF/1000G.EUR.{chrom_num} --keep ./prs_intermediate/indivs/80_percent_ids.txt --make-bed --out ./prs_intermediate/train/train_set\"\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        result = subprocess.run(plink_command, shell=True, check=True, capture_output=True, text=True)\n",
    "        print(\"PLINK output:\", result.stdout)  \n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Error occurred:\", e.stderr)  \n",
    "    \n",
    "    # Filter SNPs  \n",
    "    start, stop = cis_range(gene_name)\n",
    "    \n",
    "    print(start, stop)\n",
    "        \n",
    "    plink_command = (\n",
    "        f\".\\plink --bfile ./prs_intermediate/train/train_set --keep ./prs_intermediate/indivs/80_percent_ids.txt --chr 22 --from-bp {start} --to-bp {stop} --allow-no-sex --make-bed --out ./prs_intermediate/train/cis_gene\"\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        result = subprocess.run(plink_command, shell=True, check=True, capture_output=True, text=True)\n",
    "        print(\"PLINK output:\", result.stdout)  \n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Error occurred:\", e.stderr) \n",
    "        \n",
    "    # Linear Regression\n",
    "        \n",
    "    plink_command = (\n",
    "        f\".\\plink --bfile ./train/cis_gene --linear --pheno gene_exp_{gene_name}_{mode}.txt  --out ./prs_intermediate/train/cis_eqtl_results_train --allow-no-sex\"\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        result = subprocess.run(plink_command, shell=True, check=True, capture_output=True, text=True)\n",
    "        print(\"PLINK output:\", result.stdout)  \n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Error occurred:\", e.stderr)  \n",
    "        \n",
    "    \n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./train/train_set.log.\n",
      "Options in effect:\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --keep ./indivs/80_percent_ids.txt\n",
      "  --make-bed\n",
      "  --out ./train/train_set\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./train/train_set.nosex .\n",
      "--keep: 391 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 391 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is exactly 1.\n",
      "17489 variants and 391 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to ./train/train_set.bed + ./train/train_set.bim +\n",
      "./train/train_set.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "16637510 17665287\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./train/cis_gene.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./train/train_set\n",
      "  --chr 22\n",
      "  --from-bp 16637510\n",
      "  --keep ./indivs/80_percent_ids.txt\n",
      "  --make-bed\n",
      "  --out ./train/cis_gene\n",
      "  --to-bp 17665287\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "281 out of 17489 variants loaded from .bim file.\n",
      "391 people (0 males, 0 females, 391 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./train/cis_gene.nosex .\n",
      "--keep: 391 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 391 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "281 variants and 391 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to ./train/cis_gene.bed + ./train/cis_gene.bim +\n",
      "./train/cis_gene.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./train/cis_eqtl_results_train.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./train/cis_gene\n",
      "  --linear\n",
      "  --out ./train/cis_eqtl_results_train\n",
      "  --pheno gene_exp_CECR5_train.txt\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "281 variants loaded from .bim file.\n",
      "391 people (0 males, 0 females, 391 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./train/cis_eqtl_results_train.nosex .\n",
      "278 phenotype values present after --pheno.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 391 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "281 variants and 391 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "Writing linear model association results to\n",
      "./train/cis_eqtl_results_train.assoc.linear ... done.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "one_gene_cis_eqtl(gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bonferroni threshold: 0.00017793594306049823\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BP</th>\n",
       "      <th>A1</th>\n",
       "      <th>TEST</th>\n",
       "      <th>NMISS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>STAT</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>22</td>\n",
       "      <td>rs3827295</td>\n",
       "      <td>17600497</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>-3.696</td>\n",
       "      <td>-4.074</td>\n",
       "      <td>6.050000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>22</td>\n",
       "      <td>rs5747004</td>\n",
       "      <td>17614011</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>-4.500</td>\n",
       "      <td>-4.949</td>\n",
       "      <td>1.298000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>22</td>\n",
       "      <td>rs2302521</td>\n",
       "      <td>17624072</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>-5.077</td>\n",
       "      <td>-5.553</td>\n",
       "      <td>6.612000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>22</td>\n",
       "      <td>rs5994180</td>\n",
       "      <td>17634050</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>-5.970</td>\n",
       "      <td>-6.412</td>\n",
       "      <td>6.196000e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHR        SNP        BP A1 TEST  NMISS   BETA   STAT             P\n",
       "241   22  rs3827295  17600497  A  ADD    278 -3.696 -4.074  6.050000e-05\n",
       "251   22  rs5747004  17614011  A  ADD    278 -4.500 -4.949  1.298000e-06\n",
       "261   22  rs2302521  17624072  A  ADD    278 -5.077 -5.553  6.612000e-08\n",
       "265   22  rs5994180  17634050  G  ADD    278 -5.970 -6.412  6.196000e-10"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_results = pd.read_csv(\"./prs_intermediate/train/cis_eqtl_results_train.assoc.linear\", sep='\\s+')\n",
    "\n",
    "bonferroni_threshold = 0.05 / len(linear_results)\n",
    "print(f\"Bonferroni threshold: {bonferroni_threshold}\")\n",
    "\n",
    "# Filter significant SNPs\n",
    "significant_snps_bonferroni = linear_results[linear_results[\"P\"] < bonferroni_threshold]\n",
    "significant_snps_bonferroni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command executed successfully.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# PowerShell command to extract the 3rd and 8th columns\n",
    "powershell_command = \"\"\"\n",
    "Get-Content ./prs_intermediate/train/cis_eqtl_results_train.assoc.linear | ForEach-Object {\n",
    "    $columns = $_ -split '\\\\s+'\n",
    "    if ($columns.Length -ge 8) {\n",
    "        \"$($columns[2]) $($columns[9])\"\n",
    "    }\n",
    "} | Set-Content ./prs_intermediate/snp-pval/SNP.pvalue\n",
    "\"\"\"\n",
    "\n",
    "# Run the PowerShell command using subprocess\n",
    "try:\n",
    "    result = subprocess.run(\n",
    "        [\"powershell\", \"-Command\", powershell_command],\n",
    "        text=True,\n",
    "        capture_output=True\n",
    "    )\n",
    "    if result.returncode == 0:\n",
    "        print(\"Command executed successfully.\")\n",
    "    else:\n",
    "        print(f\"Error: {result.stderr}\")\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BP</th>\n",
       "      <th>A1</th>\n",
       "      <th>TEST</th>\n",
       "      <th>NMISS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>STAT</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>rs7287144</td>\n",
       "      <td>16886873</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>0.21910</td>\n",
       "      <td>0.49480</td>\n",
       "      <td>0.6212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>rs5748662</td>\n",
       "      <td>16892858</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>0.43760</td>\n",
       "      <td>1.03100</td>\n",
       "      <td>0.3035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>rs5994034</td>\n",
       "      <td>16894090</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>0.04723</td>\n",
       "      <td>0.07362</td>\n",
       "      <td>0.9414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>rs4010554</td>\n",
       "      <td>16894264</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>0.30030</td>\n",
       "      <td>0.69830</td>\n",
       "      <td>0.4856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>rs4010558</td>\n",
       "      <td>16896762</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>0.30030</td>\n",
       "      <td>0.69830</td>\n",
       "      <td>0.4856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>22</td>\n",
       "      <td>rs1135909</td>\n",
       "      <td>17660556</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>0.41360</td>\n",
       "      <td>0.89360</td>\n",
       "      <td>0.3723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>22</td>\n",
       "      <td>rs3171599</td>\n",
       "      <td>17660646</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>-0.34270</td>\n",
       "      <td>-0.77580</td>\n",
       "      <td>0.4385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>22</td>\n",
       "      <td>rs1139056</td>\n",
       "      <td>17661178</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>-0.34440</td>\n",
       "      <td>-0.77750</td>\n",
       "      <td>0.4376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>22</td>\n",
       "      <td>rs7285896</td>\n",
       "      <td>17661791</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>0.31080</td>\n",
       "      <td>0.66470</td>\n",
       "      <td>0.5068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>22</td>\n",
       "      <td>rs3764847</td>\n",
       "      <td>17663117</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>278</td>\n",
       "      <td>0.43660</td>\n",
       "      <td>0.99730</td>\n",
       "      <td>0.3195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>281 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHR        SNP        BP A1 TEST  NMISS     BETA     STAT       P\n",
       "0     22  rs7287144  16886873  G  ADD    278  0.21910  0.49480  0.6212\n",
       "1     22  rs5748662  16892858  A  ADD    278  0.43760  1.03100  0.3035\n",
       "2     22  rs5994034  16894090  T  ADD    278  0.04723  0.07362  0.9414\n",
       "3     22  rs4010554  16894264  A  ADD    278  0.30030  0.69830  0.4856\n",
       "4     22  rs4010558  16896762  A  ADD    278  0.30030  0.69830  0.4856\n",
       "..   ...        ...       ... ..  ...    ...      ...      ...     ...\n",
       "276   22  rs1135909  17660556  C  ADD    278  0.41360  0.89360  0.3723\n",
       "277   22  rs3171599  17660646  G  ADD    278 -0.34270 -0.77580  0.4385\n",
       "278   22  rs1139056  17661178  G  ADD    278 -0.34440 -0.77750  0.4376\n",
       "279   22  rs7285896  17661791  T  ADD    278  0.31080  0.66470  0.5068\n",
       "280   22  rs3764847  17663117  C  ADD    278  0.43660  0.99730  0.3195\n",
       "\n",
       "[281 rows x 9 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "cis_eqtl_results = pd.read_csv('cis-eQTL_results.txt', sep='\\s+')\n",
    "\n",
    "cis_eqtl_results_unique = cis_eqtl_results.drop_duplicates(subset=['snp'], keep='first')\n",
    "\n",
    "cis_eqtl_results_unique.to_csv('cis-eQTL_results_unique.txt', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>snp</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>TPTEP1</td>\n",
       "      <td>rs4911642</td>\n",
       "      <td>16504399</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.004224</td>\n",
       "      <td>0.006564</td>\n",
       "      <td>0.001209</td>\n",
       "      <td>0.520345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>CECR2</td>\n",
       "      <td>rs7287144</td>\n",
       "      <td>16886873</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.003072</td>\n",
       "      <td>0.004847</td>\n",
       "      <td>0.001173</td>\n",
       "      <td>0.526606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>CECR1</td>\n",
       "      <td>rs5748662</td>\n",
       "      <td>16892858</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.888838</td>\n",
       "      <td>0.385790</td>\n",
       "      <td>0.015284</td>\n",
       "      <td>0.021825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>IL17RA</td>\n",
       "      <td>rs5994034</td>\n",
       "      <td>16894090</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.051625</td>\n",
       "      <td>0.166538</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>0.756759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>CECR1</td>\n",
       "      <td>rs4010554</td>\n",
       "      <td>16894264</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.806934</td>\n",
       "      <td>0.391055</td>\n",
       "      <td>0.012297</td>\n",
       "      <td>0.039822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16417</th>\n",
       "      <td>22</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>rs13056621</td>\n",
       "      <td>51181759</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.506812</td>\n",
       "      <td>0.370437</td>\n",
       "      <td>0.005443</td>\n",
       "      <td>0.172164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16418</th>\n",
       "      <td>22</td>\n",
       "      <td>SHANK3</td>\n",
       "      <td>rs3865766</td>\n",
       "      <td>51186228</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000840</td>\n",
       "      <td>0.001459</td>\n",
       "      <td>0.000968</td>\n",
       "      <td>0.565229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16419</th>\n",
       "      <td>22</td>\n",
       "      <td>RABL2B</td>\n",
       "      <td>rs3888396</td>\n",
       "      <td>51211392</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.136969</td>\n",
       "      <td>0.251063</td>\n",
       "      <td>0.000870</td>\n",
       "      <td>0.585725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16420</th>\n",
       "      <td>22</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>rs2238837</td>\n",
       "      <td>51212875</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.281082</td>\n",
       "      <td>0.263877</td>\n",
       "      <td>0.003307</td>\n",
       "      <td>0.287537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16421</th>\n",
       "      <td>22</td>\n",
       "      <td>SHANK3</td>\n",
       "      <td>rs28729663</td>\n",
       "      <td>51219006</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.002186</td>\n",
       "      <td>0.002070</td>\n",
       "      <td>0.003248</td>\n",
       "      <td>0.291836</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16422 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom    gene_id         snp       pos ref alt      beta        se  \\\n",
       "0         22     TPTEP1   rs4911642  16504399   C   T -0.004224  0.006564   \n",
       "1         22      CECR2   rs7287144  16886873   G   A -0.003072  0.004847   \n",
       "2         22      CECR1   rs5748662  16892858   A   G -0.888838  0.385790   \n",
       "3         22     IL17RA   rs5994034  16894090   T   C  0.051625  0.166538   \n",
       "4         22      CECR1   rs4010554  16894264   A   C -0.806934  0.391055   \n",
       "...      ...        ...         ...       ...  ..  ..       ...       ...   \n",
       "16417     22  RPL23AP82  rs13056621  51181759   A   G  0.506812  0.370437   \n",
       "16418     22     SHANK3   rs3865766  51186228   T   C  0.000840  0.001459   \n",
       "16419     22     RABL2B   rs3888396  51211392   C   T  0.136969  0.251063   \n",
       "16420     22  RPL23AP82   rs2238837  51212875   C   A -0.281082  0.263877   \n",
       "16421     22     SHANK3  rs28729663  51219006   A   G  0.002186  0.002070   \n",
       "\n",
       "       r_squared   p_value  \n",
       "0       0.001209  0.520345  \n",
       "1       0.001173  0.526606  \n",
       "2       0.015284  0.021825  \n",
       "3       0.000281  0.756759  \n",
       "4       0.012297  0.039822  \n",
       "...          ...       ...  \n",
       "16417   0.005443  0.172164  \n",
       "16418   0.000968  0.565229  \n",
       "16419   0.000870  0.585725  \n",
       "16420   0.003307  0.287537  \n",
       "16421   0.003248  0.291836  \n",
       "\n",
       "[16422 rows x 10 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('cis-eQTL_results_unique.txt', sep='\\s+')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_prs(mode='valid', p_vals = [1, 0.05, 0.01, 0.001, 0.00001]):\n",
    "\n",
    "    \n",
    "    for p_val in p_vals:\n",
    "        clump(p_val)\n",
    "        valid_snp_idx(p_val)\n",
    "        \n",
    "        if mode == 'valid':\n",
    "            id_set='./prs_intermediate/indivs/validation_set_ids.txt'\n",
    "        \n",
    "        if mode == 'test':\n",
    "            id_set='./prs_intermediate/indivs/test_set_ids.txt'\n",
    "        \n",
    "        plink_command = (\n",
    "            f\"plink --bfile ./LDREF/1000G.EUR.{chrom_num} \"\n",
    "            f\"--keep {id_set} \"\n",
    "            f\"--score cis-eQTL_results_unique.txt 3 6 7 header \"\n",
    "            f\"--extract ./prs_intermediate/valid-snp/EUR__{p_val}.valid.snp \"\n",
    "            f\"--out ./results/prs_model_building/{mode}/EUR__{p_val}\"\n",
    "        )\n",
    "        \n",
    "        \n",
    "        try:\n",
    "            result = subprocess.run(plink_command, shell=True, check=True, capture_output=True, text=True)\n",
    "            print(\"PLINK output:\", result.stdout)  \n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(\"Error occurred:\", e.stderr)  \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./clumps/EUR__1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --clump cis-eqtl_results.txt\n",
      "  --clump-field p_value\n",
      "  --clump-kb 250\n",
      "  --clump-p1 1\n",
      "  --clump-r2 0.1\n",
      "  --clump-snp-field snp\n",
      "  --out ./clumps/EUR__1\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./clumps/EUR__1.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 489 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "17489 variants and 489 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--clump: 2130 clumps formed from 16422 top variants.\n",
      "Results written to ./clumps/EUR__1.clumped .\n",
      "\n",
      "Index SNP IDs extracted.\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./valid/EUR__1.log.\n",
      "Options in effect:\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --extract ./valid-snp/EUR__1.valid.snp\n",
      "  --keep ./indivs/validation_set_ids.txt\n",
      "  --out ./valid/EUR__1\n",
      "  --score cis-eQTL_results_unique.txt 3 6 7 header\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./valid/EUR__1.nosex .\n",
      "--extract: 2130 variants remaining.\n",
      "--keep: 49 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 49 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is in [0.9999995, 1).\n",
      "2130 variants and 49 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 2130 valid predictors loaded.\n",
      "--score: Results written to ./valid/EUR__1.profile .\n",
      "\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./clumps/EUR__0.05.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --clump cis-eqtl_results.txt\n",
      "  --clump-field p_value\n",
      "  --clump-kb 250\n",
      "  --clump-p1 0.05\n",
      "  --clump-r2 0.1\n",
      "  --clump-snp-field snp\n",
      "  --out ./clumps/EUR__0.05\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./clumps/EUR__0.05.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 489 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "17489 variants and 489 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--clump: 1118 clumps formed from 5922 top variants.\n",
      "Results written to ./clumps/EUR__0.05.clumped .\n",
      "\n",
      "Index SNP IDs extracted.\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./valid/EUR__0.05.log.\n",
      "Options in effect:\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --extract ./valid-snp/EUR__0.05.valid.snp\n",
      "  --keep ./indivs/validation_set_ids.txt\n",
      "  --out ./valid/EUR__0.05\n",
      "  --score cis-eQTL_results_unique.txt 3 6 7 header\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./valid/EUR__0.05.nosex .\n",
      "--extract: 1118 variants remaining.\n",
      "--keep: 49 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 49 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is in [0.9999995, 1).\n",
      "1118 variants and 49 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 1118 valid predictors loaded.\n",
      "--score: Results written to ./valid/EUR__0.05.profile .\n",
      "\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./clumps/EUR__0.01.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --clump cis-eqtl_results.txt\n",
      "  --clump-field p_value\n",
      "  --clump-kb 250\n",
      "  --clump-p1 0.01\n",
      "  --clump-r2 0.1\n",
      "  --clump-snp-field snp\n",
      "  --out ./clumps/EUR__0.01\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./clumps/EUR__0.01.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 489 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "17489 variants and 489 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--clump: 450 clumps formed from 1502 top variants.\n",
      "Results written to ./clumps/EUR__0.01.clumped .\n",
      "\n",
      "Index SNP IDs extracted.\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./valid/EUR__0.01.log.\n",
      "Options in effect:\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --extract ./valid-snp/EUR__0.01.valid.snp\n",
      "  --keep ./indivs/validation_set_ids.txt\n",
      "  --out ./valid/EUR__0.01\n",
      "  --score cis-eQTL_results_unique.txt 3 6 7 header\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./valid/EUR__0.01.nosex .\n",
      "--extract: 450 variants remaining.\n",
      "--keep: 49 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 49 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is in [0.9999995, 1).\n",
      "450 variants and 49 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 450 valid predictors loaded.\n",
      "--score: Results written to ./valid/EUR__0.01.profile .\n",
      "\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./clumps/EUR__0.001.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --clump cis-eqtl_results.txt\n",
      "  --clump-field p_value\n",
      "  --clump-kb 250\n",
      "  --clump-p1 0.001\n",
      "  --clump-r2 0.1\n",
      "  --clump-snp-field snp\n",
      "  --out ./clumps/EUR__0.001\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./clumps/EUR__0.001.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 489 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "17489 variants and 489 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--clump: 87 clumps formed from 173 top variants.\n",
      "Results written to ./clumps/EUR__0.001.clumped .\n",
      "\n",
      "Index SNP IDs extracted.\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./valid/EUR__0.001.log.\n",
      "Options in effect:\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --extract ./valid-snp/EUR__0.001.valid.snp\n",
      "  --keep ./indivs/validation_set_ids.txt\n",
      "  --out ./valid/EUR__0.001\n",
      "  --score cis-eQTL_results_unique.txt 3 6 7 header\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./valid/EUR__0.001.nosex .\n",
      "--extract: 87 variants remaining.\n",
      "--keep: 49 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 49 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is in [0.9999995, 1).\n",
      "87 variants and 49 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 87 valid predictors loaded.\n",
      "--score: Results written to ./valid/EUR__0.001.profile .\n",
      "\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./clumps/EUR__1e-05.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --clump cis-eqtl_results.txt\n",
      "  --clump-field p_value\n",
      "  --clump-kb 250\n",
      "  --clump-p1 1e-05\n",
      "  --clump-r2 0.1\n",
      "  --clump-snp-field snp\n",
      "  --out ./clumps/EUR__1e-05\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./clumps/EUR__1e-05.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 489 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "17489 variants and 489 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--clump: 6 clumps formed from 10 top variants.\n",
      "Results written to ./clumps/EUR__1e-05.clumped .\n",
      "\n",
      "Index SNP IDs extracted.\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./valid/EUR__1e-05.log.\n",
      "Options in effect:\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --extract ./valid-snp/EUR__1e-05.valid.snp\n",
      "  --keep ./indivs/validation_set_ids.txt\n",
      "  --out ./valid/EUR__1e-05\n",
      "  --score cis-eQTL_results_unique.txt 3 6 7 header\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./valid/EUR__1e-05.nosex .\n",
      "--extract: 6 variants remaining.\n",
      "--keep: 49 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 49 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is in [0.9999995, 1).\n",
      "6 variants and 49 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 6 valid predictors loaded.\n",
      "--score: Results written to ./valid/EUR__1e-05.profile .\n",
      "\n"
     ]
    }
   ],
   "source": [
    "generate_prs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best threshold: 1e-05, R2: 0.10916119452856231\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Load true values\n",
    "gene_exp_txt(gene_name=gene_name, mode='valid')\n",
    "true_values = pd.read_csv(f\"gene_exp_{gene_name}_valid.txt\", sep='\\s+')\n",
    "true_values.columns = [\"FID\", \"IID\", \"TrueValue\"]\n",
    "\n",
    "# Iterate through thresholds\n",
    "thresholds = [1, 0.05, 0.01, 0.001, 0.00001]\n",
    "results = []\n",
    "\n",
    "for threshold in thresholds:\n",
    "    prs = pd.read_csv(f\"./results/prs_model_building/valid/EUR__{threshold}.profile\", sep='\\s+')\n",
    "    merged = pd.merge(prs, true_values, on=[\"FID\", \"IID\"])\n",
    "    r2 = (merged.corr('pearson', numeric_only=True)['SCORE']['TrueValue'])**2 \n",
    "    results.append({\"Threshold\": threshold, \"R2\": r2})\n",
    "\n",
    "# Find best threshold\n",
    "best_result = max(results, key=lambda x: x[\"R2\"])\n",
    "print(f\"Best threshold: {best_result['Threshold']}, R2: {best_result['R2']}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'Threshold': 1, 'R2': 0.016876544273791837},\n",
       " {'Threshold': 0.05, 'R2': 0.0020655018565138026},\n",
       " {'Threshold': 0.01, 'R2': 0.011898179674400055},\n",
       " {'Threshold': 0.001, 'R2': 0.0061382567678518385},\n",
       " {'Threshold': 1e-05, 'R2': 0.10916119452856231}]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./clumps/EUR__1e-05.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --clump cis-eqtl_results.txt\n",
      "  --clump-field p_value\n",
      "  --clump-kb 250\n",
      "  --clump-p1 1e-05\n",
      "  --clump-r2 0.1\n",
      "  --clump-snp-field snp\n",
      "  --out ./clumps/EUR__1e-05\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./clumps/EUR__1e-05.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 489 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is exactly 1.\n",
      "17489 variants and 489 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--clump: 6 clumps formed from 10 top variants.\n",
      "Results written to ./clumps/EUR__1e-05.clumped .\n",
      "\n",
      "Index SNP IDs extracted.\n",
      "PLINK output: PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./test/EUR__1e-05.log.\n",
      "Options in effect:\n",
      "  --bfile ./LDREF/1000G.EUR.22\n",
      "  --extract ./valid-snp/EUR__1e-05.valid.snp\n",
      "  --keep ./indivs/test_set_ids.txt\n",
      "  --out ./test/EUR__1e-05\n",
      "  --score cis-eQTL_results_unique.txt 3 6 7 header\n",
      "\n",
      "16075 MB RAM detected; reserving 8037 MB for main workspace.\n",
      "17489 variants loaded from .bim file.\n",
      "489 people (0 males, 0 females, 489 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to ./test/EUR__1e-05.nosex .\n",
      "--extract: 6 variants remaining.\n",
      "--keep: 49 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 49 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is in [0.9999995, 1).\n",
      "6 variants and 49 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 6 valid predictors loaded.\n",
      "--score: Results written to ./test/EUR__1e-05.profile .\n",
      "\n"
     ]
    }
   ],
   "source": [
    "generate_prs(mode='test', p_vals=[best_result['Threshold']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.06546655770727856"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "gene_exp_txt(gene_name=gene_name, mode='test')\n",
    "true_values = pd.read_csv(f\"gene_exp_{gene_name}_test.txt\", sep='\\s+')\n",
    "true_values.columns = [\"FID\", \"IID\", \"TrueValue\"]\n",
    "\n",
    "prs = pd.read_csv(f\"./results/prs_model_building/test/EUR__{best_result['Threshold']}.profile\", sep='\\s+')\n",
    "merged = pd.merge(prs, true_values, on=[\"FID\", \"IID\"])\n",
    "r2 = (merged.corr('pearson', numeric_only=True)['SCORE']['TrueValue'])**2 \n",
    "r2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlAAAAHHCAYAAABwaWYjAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAA9hAAAPYQGoP6dpAABVz0lEQVR4nO3deVxU9f4/8NeZGWZYhmFRRkARBPfcIXHHRAUtr5blcumXqKG5tGhWeO83rbylXS3Nvt5sdUuvpmld7appLqUhLqWmIYJKKoKYyL7N8vn9gczXkWE5Cgzg6/l4zCPnrO/z4cC8OucznyMJIQSIiIiIqNoU9i6AiIiIqKFhgCIiIiKSiQGKiIiISCYGKCIiIiKZGKCIiIiIZGKAIiIiIpKJAYqIiIhIJgYoIiIiIpkYoIiIiIhkYoAiIqJ67cCBA5AkCQcOHLB3KUQWDFBENWD16tWQJMnycnR0RNu2bTFz5kxcv37dslzZB0HZS6lUQq/X48knn0RCQoLNbW/fvh1hYWHQ6/VwdnZGYGAgxowZg127dtXV4ZUTHR1tdRwVvaKjo+1W450uXbqEmTNnom3btnB2doazszM6duyIGTNm4PTp0/Yur1J3nzN3vzZu3GjvEokeSCp7F0DUmLz11lto1aoVioqKcOjQIXz00Uf473//izNnzsDZ2dmy3AsvvICHH34YBoMBp0+fxsqVK3HgwAGcOXMG3t7eluWWLFmCV155BWFhYZg7dy6cnZ2RnJyMvXv3YuPGjYiMjLTHYWLq1KkYPHiw5f2lS5cwb948TJkyBf3797dMDwoKskd5Vnbs2IGxY8dCpVIhKioKXbt2hUKhwLlz57B161Z89NFHuHTpEvz9/e1daqXKzpm79e7d2w7V1K0BAwagsLAQarXa3qUQ/R9BRPdt1apVAoA4duyY1fTZs2cLAGLDhg1CCCH2798vAIjNmzdbLffRRx8JAOLdd9+1TDMYDEKn04khQ4bY3Of169dr+Cju3bFjxwQAsWrVqkqXy8vLq5uCbktOThYuLi6iQ4cO4tq1a+XmGwwG8cEHH4jLly/XaV1yVHTO2Etd/wyJ6ivewiOqRYMGDQJQeoWmMmVXbS5cuGCZ9ueffyInJwd9+/a1uY5er690m506dcIjjzxSbrrZbEbz5s3x5JNPWqZt3LgRwcHBcHV1hU6nQ+fOnfHBBx9Uuv2qlN3WPHjwIKZPnw69Xo8WLVoAKL0FGBAQUG6dN954A5IklZv+5ZdfIjg4GE5OTvD09MS4ceNw5cqVKmv45z//ifz8fKxatQo+Pj7l5qtUKrzwwgvw8/Ozmn7u3Dk8+eST8PT0hKOjI0JCQvCf//zH5vEdPnwYs2fPhpeXF1xcXPD444/jxo0b5fa1c+dO9O/fHy4uLnB1dcWjjz6Ks2fPVnkM1bVq1SpIkoQvvvjCavo777wDSZLw3//+FwCQkpICSZKwZMkSLF26FP7+/nByckJYWBjOnDljtW50dDS0Wi0uXLiA4cOHw9XVFVFRUQBKz6Nly5bhoYcegqOjI5o1a4apU6fi1q1bVts4fvw4IiIi0LRpUzg5OaFVq1aYNGmS1TJVnX8V9YHavHmz5bxo2rQpnn76aaSmpto8htTUVIwaNQparRZeXl6YM2cOTCaT/IYmuo0BiqgWlQWiJk2aVLpcSkoKAMDDw8MyTa/Xw8nJCdu3b0dmZqbsfY8dOxY//vgj0tPTraYfOnQI165dw7hx4wAAe/bswfjx4+Hh4YF3330XixYtwsCBA3H48GHZ+7Rl+vTp+P333zFv3jzExsbKXv/tt9/GM888gzZt2uD999/HSy+9hB9++AEDBgxAVlZWpevu2LEDrVu3RmhoaLX3d/bsWfTq1QsJCQmIjY3Fe++9BxcXF4waNQrbtm0rt/zzzz+PU6dOYf78+Zg2bRq2b9+OmTNnWi2zbt06PProo9BqtXj33Xfx+uuv4/fff0e/fv0sP/uq5Obm4s8//yz3EkIAACZOnIjHHnsMs2fPtoTL3377DW+++SYmT56M4cOHW21v7dq1WL58OWbMmIG5c+fizJkzGDRokFWfPQAwGo2IiIiAXq/HkiVLMHr0aAClt3FfeeUV9O3bFx988AEmTpyI9evXIyIiAgaDAQCQkZGBoUOHIiUlBbGxsfjwww8RFRWFI0eOWLZ/r+ff6tWrMWbMGCiVSixcuBAxMTHYunUr+vXrV+68MJlMiIiIQJMmTbBkyRKEhYXhvffewyeffFKttieyyd6XwIgag7JbeHv37hU3btwQV65cERs3bhRNmjQRTk5O4urVq0KI/7sd88UXX4gbN26Ia9euiV27donWrVsLSZLE0aNHrbY7b948AUC4uLiIYcOGibffflucOHGiWjUlJiYKAOLDDz+0mj59+nSh1WpFQUGBEEKIF198Ueh0OmE0Gu/5+G3dwitrk379+pXb9oQJE4S/v3+57cyfP1/c+WcpJSVFKJVK8fbbb1st99tvvwmVSlVu+p2ys7MFADFq1Khy827duiVu3LhheZW1hRBChIeHi86dO4uioiLLNLPZLPr06SPatGlT7vgGDx4szGazZfqsWbOEUqkUWVlZQgghcnNzhbu7u4iJibGqIT09Xbi5uZWbfreyc6aiV1pammXZtLQ04enpKYYMGSKKi4tF9+7dRcuWLUV2drZlmUuXLgkAVuelEELEx8cLAGLWrFmWaRMmTBAARGxsrFVNP/30kwAg1q9fbzV9165dVtO3bdtm89b2napz/pW1wf79+4UQQpSUlAi9Xi86deokCgsLLcvt2LFDABDz5s0rdwxvvfWW1Ta7d+8ugoODK9wnUVV4BYqoBg0ePBheXl7w8/PDuHHjoNVqsW3bNjRv3txquUmTJsHLywu+vr6IjIxEdnY21q1bV66T8JtvvokNGzage/fu2L17N/7+978jODgYPXr0qPBbe2Xatm2Lbt26YdOmTZZpJpMJW7ZswYgRI+Dk5AQAcHd3R35+Pvbs2VNDrWAtJiYGSqXyntbdunUrzGYzxowZY3XVxdvbG23atMH+/fsrXDcnJwcAoNVqy80bOHAgvLy8LK8VK1YAADIzM7Fv3z6MGTPG6orPzZs3ERERgaSkpHK3iKZMmWJ127F///4wmUz4448/AJReYcnKysL48eOtjkGpVCI0NLTSY7jTvHnzsGfPnnIvT09PyzLe3t5YsWIF9uzZg/79++PkyZP44osvoNPpym1v1KhRVudlz549ERoaarnVd6dp06ZZvd+8eTPc3NwwZMgQq2MKDg6GVqu1HJO7uzuA0iuBZVel7nYv59/x48eRkZGB6dOnw9HR0TL90UcfRfv27fHdd9+VW+e5556zet+/f39cvHix2vskuhu/hUdUg1asWIG2bdtCpVKhWbNmaNeuHRSK8v+fMm/ePPTv3x95eXnYtm0bNm7caHM5ABg/fjzGjx+PnJwcxMfHY/Xq1diwYQNGjBiBM2fOWH2A3G3s2LH429/+htTUVDRv3hwHDhxARkYGxo4da1lm+vTp+OqrrzBs2DA0b94cQ4cOxZgxY2rsG36tWrW653WTkpIghECbNm1szndwcKhwXVdXVwBAXl5euXkff/wxcnNzcf36dTz99NOW6cnJyRBC4PXXX8frr79uc7sZGRlWwaNly5ZW88tuw5b1BUpKSgLwf/3h7mYr3NjSuXNnq28+VmTcuHH48ssv8d1332HKlCkIDw+3uZytNm3bti2++uorq2kqlcrSd61MUlISsrOzK+yHl5GRAQAICwvD6NGj8eabb2Lp0qUYOHAgRo0ahb/+9a/QaDQA7u38Kwun7dq1Kzevffv2OHTokNU0R0dHeHl5WU3z8PAo11+LSA4GKKIa1LNnT4SEhFS53J0fhqNGjUJBQQFiYmLQr1+/ch2ay+h0OgwZMgRDhgyBg4MD1qxZg/j4eISFhVW4n7Fjx2Lu3LnYvHkzXnrpJXz11Vdwc3Oz+nDS6/U4efIkdu/ejZ07d2Lnzp1YtWoVnnnmGaxZs0ZmC5RXdqXrTrY6igMo16nXbDZDkiTs3LnT5lUsW1eXyri5ucHHx6dcx2gAlj5Rd/c/MpvNAIA5c+YgIiLC5nZbt25t9b6iq2vidt+ksm2uW7fOaoiKMipVzf4ZvnnzJo4fPw4A+P3332E2mysM59Wh0WjKrW82m6HX67F+/Xqb65SFFUmSsGXLFhw5cgTbt2/H7t27MWnSJLz33ns4cuQItFptrZ9/QMU/I6L7wQBFVA8sWrQI27Ztw9tvv42VK1dWuXxISAjWrFmDtLS0Spdr1aoVevbsiU2bNmHmzJnYunUrRo0aZfm//zJqtRojRozAiBEjYDabMX36dHz88cd4/fXXywWGmuDh4WGzA3jZlYUyQUFBEEKgVatWaNu2rez9PProo/jss89w9OhR9OzZs8rlAwMDAZRe2arO1Z7qKBsLS6/X19g2KzNjxgzk5uZi4cKFmDt3LpYtW4bZs2eXW67sytidzp8/b/PbkXcLCgrC3r170bdvX5sB+W69evVCr1698Pbbb2PDhg2IiorCxo0b8eyzzwKQf/6VjdmVmJhY7speYmJivR/TixoH9oEiqgeCgoIwevRorF692vKtuYKCAsTFxdlcfufOnQBs38K429ixY3HkyBF88cUX+PPPP61u3wGlVyzupFAo0KVLFwBAcXGx7GOpjqCgIGRnZ1uNAp6WllbuW25PPPEElEol3nzzTcsVnTJCiHK13+3VV1+Fs7MzJk2aVO7bZWXbuJNer8fAgQPx8ccf2wyntoYnqEpERAR0Oh3eeecdm/2A7mWbFdmyZQs2bdqERYsWITY2FuPGjcP//M//4Pz58+WW/eabb6z6cx09ehTx8fEYNmxYlfsZM2YMTCYTFixYUG6e0Wi0hONbt26Va+Nu3boB+L9z617Ov5CQEOj1eqxcudJqmZ07dyIhIQGPPvpolcdAdL94BYqonnjllVfw1VdfYdmyZVi0aBEKCgrQp08f9OrVC5GRkfDz80NWVha++eYb/PTTTxg1ahS6d+9e5XbHjBmDOXPmYM6cOfD09Cx3FeTZZ59FZmYmBg0ahBYtWuCPP/7Ahx9+iG7duqFDhw61cqzjxo3Da6+9hscffxwvvPACCgoK8NFHH6Ft27b45ZdfLMsFBQXhH//4B+bOnYuUlBSMGjUKrq6uuHTpErZt24YpU6Zgzpw5Fe6nTZs22LBhA8aPH4927dpZRiIXQuDSpUvYsGEDFAqFVR+fFStWoF+/fujcuTNiYmIQGBiI69evIy4uDlevXsWpU6dkHatOp8NHH32E//f//h969OiBcePGwcvLC5cvX8Z3332Hvn374n//93+r3M5PP/2EoqKictO7dOmCLl26ICMjA9OmTcMjjzxiGUbhf//3f7F//35ER0fj0KFDVrfiWrdujX79+mHatGkoLi7GsmXL0KRJE7z66qtV1hIWFoapU6di4cKFOHnyJIYOHQoHBwckJSVh8+bN+OCDD/Dkk09izZo1+Ne//oXHH38cQUFByM3NxaeffgqdTmcZVuFezj8HBwe8++67mDhxIsLCwjB+/Hhcv34dH3zwAQICAjBr1qwqj4Hovtnt+39EjUhFI5HfrapRpQcOHCh0Op3IysoSBoNBfPrpp2LUqFHC399faDQa4ezsLLp37y4WL14siouLq11f3759BQDx7LPPlpu3ZcsWMXToUKHX64VarRYtW7YUU6dOtfp6fFUqG8agojb5/vvvRadOnYRarRbt2rUTX375ZblhDMp8/fXXol+/fsLFxUW4uLiI9u3bixkzZojExMRq1ZecnCymTZsmWrduLRwdHYWTk5No3769eO6558TJkyfLLX/hwgXxzDPPCG9vb+Hg4CCaN28uHnvsMbFly5Yqj+/ur9zfOT0iIkK4ubkJR0dHERQUJKKjo8Xx48crrb2qYQzmz58vhBDiiSeeEK6uriIlJcVq/W+//dZqlPuyYQwWL14s3nvvPeHn5yc0Go3o37+/OHXqlNW6EyZMEC4uLhXW9sknn4jg4GDh5OQkXF1dRefOncWrr75qGfX9l19+EePHjxctW7YUGo1G6PV68dhjj1kdc3XOv4radNOmTaJ79+5Co9EIT09PERUVZTU0Q2XHUNG5RlRdkhB3XV8lIqJGKyUlBa1atcLixYsrvXpHRJVjHygiIiIimRigiIiIiGRigCIiIiKSiX2giIiIiGTiFSgiIiIimRigiIiIiGTiQJo1wGw249q1a3B1da3wGV9ERERUvwghkJubC19fX9nPjGSAqgHXrl2r8AGwREREVL9duXLF6okE1cEAVQNcXV0BlP4AdDqdnashIiKi6sjJyYGfn5/lc1wOBqgaUHbbTqfTMUARERE1MPfS/YadyImIiIhkYoAiIiIikokBioiIiEgmBigiIiIimRigiIiIiGRigCIiIiKSiQGKiIiISCYGKCIiIiKZGKCIiIiIZOJI5ERERFQls1kgNasQ+SVGuKhVaO7uBIVC/gjejQUDFBEREVUqOSMXu89cx4UbeSgymuCoUiLIS4uITs3QWi//OXKNAQMUERERVSg5IxerDqcgM78EPm6OcFY7oaDEiDPXsnEtuxAT+wY8kCGKfaCIiIjIJrNZYPeZ68jML0EbvRaujg5QKiS4OjqgjV6LzPwSfH/2OsxmYe9S6xwDFBEREdmUmlWICzfy4OPmCEmy7u8kSRJ83ByRnJGH1KxCO1VoPwxQREREZFN+iRFFRhOc1bZ7/DiplSg2mpBfYqzjyuyPAYqIiIhsclGr4KhSoqCCgFRYYoJGpYRLBQGrMWOAIiIiIpuauzshyEuLtOwiCGHdz0kIgbTsIrTWa9Hc3clOFdoPAxQRERHZpFBIiOjUDJ4uaiRl5CG3yACj2YzcIgOSMvLg6aLG0IeaPZDjQTFAERERUYVa610xsW8AOvm6IavAgJQ/85FVYEDn5m4P7BAGAMeBIiIioiq01rsicKCWI5HfgQGKiIiIqqRQSPDzdLZ3GfUGb+ERERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcmksncBRERE1PiZzQKpWYXILzHCRa1Cc3cnKBSSvcu6ZwxQREREVKuSM3Kx+8x1XLiRhyKjCY4qJYK8tIjo1Ayt9a72Lu+eMEARERFRrUnOyMWqwynIzC+Bj5sjnNVOKCgx4sy1bFzLLsTEvgENMkSxDxQRERHVCrNZYPeZ68jML0EbvRaujg5QKiS4OjqgjV6LzPwSfH/2OsxmYe9SZWOAIiIiolqRmlWICzfy4OPmCEmy7u8kSRJ83ByRnJGH1KxCO1V47xigiIiIqFbklxhRZDTBWW27x5CTWoliown5JcY6ruz+MUARERFRrXBRq+CoUqKggoBUWGKCRqWESwUBqz5jgCIiIqJa0dzdCUFeWqRlF0EI635OQgikZRehtV6L5u5Odqrw3jFAERERUa1QKCREdGoGTxc1kjLykFtkgNFsRm6RAUkZefB0UWPoQ80a5HhQDSZAZWZmIioqCjqdDu7u7pg8eTLy8vIqXWfgwIGQJMnq9dxzz1ktc+zYMYSHh8Pd3R0eHh6IiIjAqVOnavNQiIiIHhit9a6Y2DcAnXzdkFVgQMqf+cgqMKBzc7cGO4QB0IDGgYqKikJaWhr27NkDg8GAiRMnYsqUKdiwYUOl68XExOCtt96yvHd2drb8Oy8vD5GRkfjLX/6Cf/3rXzAajZg/fz4iIiJw5coVODg41NrxEBERPSha610ROFDLkcjrWkJCAnbt2oVjx44hJCQEAPDhhx9i+PDhWLJkCXx9fStc19nZGd7e3jbnnTt3DpmZmXjrrbfg5+cHAJg/fz66dOmCP/74A61bt675gyEiInoAKRQS/Dydq16wgWgQt/Di4uLg7u5uCU8AMHjwYCgUCsTHx1e67vr169G0aVN06tQJc+fORUFBgWVeu3bt0KRJE3z++ecoKSlBYWEhPv/8c3To0AEBAQEVbrO4uBg5OTlWLyIiInpwNIgrUOnp6dDr9VbTVCoVPD09kZ6eXuF6f/3rX+Hv7w9fX1+cPn0ar732GhITE7F161YAgKurKw4cOIBRo0ZhwYIFAIA2bdpg9+7dUKkqbpqFCxfizTffrIEjIyIioobIrlegYmNjy3Xyvvt17ty5e97+lClTEBERgc6dOyMqKgpr167Ftm3bcOHCBQBAYWEhJk+ejL59++LIkSM4fPgwOnXqhEcffRSFhRWPijp37lxkZ2dbXleuXLnnGomIiKjhsesVqJdffhnR0dGVLhMYGAhvb29kZGRYTTcajcjMzKywf5MtoaGhAIDk5GQEBQVhw4YNSElJQVxcHBSK0iy5YcMGeHh44Ntvv8W4ceNsbkej0UCj0VR7v0RERNS42DVAeXl5wcvLq8rlevfujaysLJw4cQLBwcEAgH379sFsNltCUXWcPHkSAODj4wMAKCgogEKhsHo+T9l7s9ks40iIiIjoQdIgOpF36NABkZGRiImJwdGjR3H48GHMnDkT48aNs3wDLzU1Fe3bt8fRo0cBABcuXMCCBQtw4sQJpKSk4D//+Q+eeeYZDBgwAF26dAEADBkyBLdu3cKMGTOQkJCAs2fPYuLEiVCpVHjkkUfsdrxERERUvzWIAAWUfpuuffv2CA8Px/Dhw9GvXz988sknlvkGgwGJiYmWb9mp1Wrs3bsXQ4cORfv27fHyyy9j9OjR2L59u2Wd9u3bY/v27Th9+jR69+6N/v3749q1a9i1a5flKhURERHR3SRx98NpSLacnBy4ubkhOzsbOp3O3uUQERFRNdzP53eDuQJFREREVF8wQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCRTgwlQmZmZiIqKgk6ng7u7OyZPnoy8vLwq14uLi8OgQYPg4uICnU6HAQMGoLCw8L63S0RERA+uBhOgoqKicPbsWezZswc7duzAjz/+iClTplS6TlxcHCIjIzF06FAcPXoUx44dw8yZM6FQ/N9h38t2iYiI6MEmCSGEvYuoSkJCAjp27Ihjx44hJCQEALBr1y4MHz4cV69eha+vr831evXqhSFDhmDBggU1ut275eTkwM3NDdnZ2dDpdPdwhERERFTX7ufzu0FcgYqLi4O7u7sl5ADA4MGDoVAoEB8fb3OdjIwMxMfHQ6/Xo0+fPmjWrBnCwsJw6NCh+9ouABQXFyMnJ8fqRURERA+OBhGg0tPTodfrraapVCp4enoiPT3d5joXL14EALzxxhuIiYnBrl270KNHD4SHhyMpKemetwsACxcuhJubm+Xl5+d3P4dHREREDYxdA1RsbCwkSar0de7cuXvattlsBgBMnToVEydORPfu3bF06VK0a9cOX3zxxX3VPXfuXGRnZ1teV65cua/tERERUcOisufOX375ZURHR1e6TGBgILy9vZGRkWE13Wg0IjMzE97e3jbX8/HxAQB07NjRanqHDh1w+fJlALin7QKARqOBRqOptG4iIiJqvOwaoLy8vODl5VXlcr1790ZWVhZOnDiB4OBgAMC+fftgNpsRGhpqc52AgAD4+voiMTHRavr58+cxbNiwe94uERERUYPoA9WhQwdERkYiJiYGR48exeHDhzFz5kyMGzfO8k251NRUtG/fHkePHgUASJKEV155BcuXL8eWLVuQnJyM119/HefOncPkyZOrvV0iIiKiu9n1CpQc69evx8yZMxEeHg6FQoHRo0dj+fLllvkGgwGJiYkoKCiwTHvppZdQVFSEWbNmITMzE127dsWePXsQFBRU7e0SERER3a1BjANV33EcKCIiooan0Y8DRURERFSfMEARERERycQARURERCRTtQPUn3/+iT/++MNq2tmzZzFx4kSMGTMGGzZsqPHiiIiIiOqjageo559/3urbaRkZGejfvz+OHTuG4uJiREdHY926dbVSJBEREVF9Uu0AdeTIEfzlL3+xvF+7di08PT1x8uRJfPvtt3jnnXewYsWKWimSiIiIqD6pdoBKT09HQECA5f2+ffvwxBNPQKUqHUrqL3/5i+UhvURERESNWbUDlE6nQ1ZWluX90aNHrR53IkkSiouLa7Q4IiIiovqo2gGqV69eWL58OcxmM7Zs2YLc3FwMGjTIMv/8+fPw8/OrlSKJiIiI6pNqP8plwYIFCA8Px5dffgmj0Yi//e1v8PDwsMzfuHEjwsLCaqVIIiIiovqk2gGqS5cuSEhIwOHDh+Ht7W11+w4Axo0bh44dO9Z4gURERET1TY0+Cy81NRXNmzevqc01GHwWHhERUcNj92fhpaen4/nnn0ebNm1qYnNERERE9Vq1A9StW7cwfvx4NG3aFL6+vpYO5fPmzUNgYCCOHTuGVatW1WatRERERPVCtftAxcbG4ueff0Z0dDR2796NWbNmYdeuXVAoFNi3bx969epVm3USERER1RvVvgK1c+dOrFq1CkuWLMH27dshhEC3bt2wY8cOhiciIiJ6oFQ7QF27dg0dOnQAAAQEBMDR0RFPP/10rRVGREREVF9VO0AJISyPbQEApVIJJyenWimKiIiIqD6rdh8oIQTCw8MtIaqwsBAjRoyAWq22Wu6XX36p2QqJiIiI6plqB6j58+dbvR85cmSNF0NERETUENToQJoPKg6kSURE1PDcz+d3ta9AAcCRI0ewfft2lJSUIDw8HJGRkbJ2RkRERNQYVDtAbdmyBWPHjoWTkxMcHBzw/vvv491338WcOXNqsz4iIiKieqfa38JbuHAhYmJikJ2djVu3buEf//gH3nnnndqsjYiIiKheqnYfKK1Wi5MnT6J169YAgJKSEri4uCA1NRV6vb5Wi6zv2AeKiIio4amThwkXFBRYbVytVsPR0RF5eXmydkhERETU0MnqRP7ZZ59Bq9Va3huNRqxevRpNmza1THvhhRdqrjoiIiKieqjat/ACAgIgSVLlG5MkXLx4sUYKa0h4C4+IiKjhqZNhDFJSUuTWRURERNQoVbsPFBERERGVYoAiIiIikokBioiIiEgmBigiIiIimRigiIiIiGSq1rfwcnJyqr1Bfo2fiIiIGrtqBSh3d/cqx4AqYzKZ7qsgIiIiovquWgFq//79ln+npKQgNjYW0dHR6N27NwAgLi4Oa9aswcKFC2unSiIiIqJ6pNojkZcJDw/Hs88+i/Hjx1tN37BhAz755BMcOHCgJutrEDgSORERUcNTJw8TLhMXF4eQkJBy00NCQnD06FG5myMiIiJqcGQHKD8/P3z66aflpn/22Wfw8/OrkaKIiIiI6rNqPwuvzNKlSzF69Gjs3LkToaGhAICjR48iKSkJX3/9dY0XSERERFTfyL4CNXz4cJw/fx4jRoxAZmYmMjMzMWLECJw/fx7Dhw+vjRqJiIiI6hXZncipPHYiJyIianjqtBM5APz00094+umn0adPH6SmpgIA1q1bh0OHDt3L5oiIiIgaFNkB6uuvv0ZERAScnJzwyy+/oLi4GACQnZ2Nd955p8YLJCIiIqpvZAeof/zjH1i5ciU+/fRTODg4WKb37dsXv/zyS40WR0RERFQfyQ5QiYmJGDBgQLnpbm5uyMrKqomaiIiIiOo12QHK29sbycnJ5aYfOnQIgYGBNVIUERERUX0mO0DFxMTgxRdfRHx8PCRJwrVr17B+/XrMmTMH06ZNq40aiYiIiOoV2QNpxsbGwmw2Izw8HAUFBRgwYAA0Gg3mzJmD559/vjZqJCIiIqpX7nkcqJKSEiQnJyMvLw8dO3aEVqut6doaDI4DRURE1PDU6ThQkyZNQm5uLtRqNTp27IiePXtCq9UiPz8fkyZNkrs5IiIiogZHdoBas2YNCgsLy00vLCzE2rVra6QoIiIiovqs2n2gcnJyIISAEAK5ublwdHS0zDOZTPjvf/8LvV5fK0USERER1SfVDlDu7u6QJAmSJKFt27bl5kuShDfffLNGiyMiIiKqj6odoPbv3w8hBAYNGoSvv/4anp6elnlqtRr+/v7w9fWtlSKJiIiI6pNqB6iwsDAAwKVLl9CyZUtIklRrRRERERHVZ7I7ke/btw9btmwpN33z5s1Ys2ZNjRRFREREVJ/JDlALFy5E06ZNy03X6/V45513aqQoIiIiovpMdoC6fPkyWrVqVW66v78/Ll++XCNF2ZKZmYmoqCjodDq4u7tj8uTJyMvLq3K9uLg4DBo0CC4uLtDpdBgwYIBlGIaUlBRMnjwZrVq1gpOTE4KCgjB//nyUlJTU2nEQERFRwyf7US56vR6nT59GQECA1fRTp06hSZMmNVVXOVFRUUhLS8OePXtgMBgwceJETJkyBRs2bKhwnbi4OERGRmLu3Ln48MMPoVKpcOrUKSgUpbnx3LlzMJvN+Pjjj9G6dWucOXMGMTExyM/Px5IlS2rtWIiIiKhhk/0ol9deew2bNm3CqlWrMGDAAADAwYMHMWnSJDz55JO1EjwSEhLQsWNHHDt2DCEhIQCAXbt2Yfjw4bh69WqF3/7r1asXhgwZggULFlR7X4sXL8ZHH32EixcvVnsdPsqFiIio4anTR7ksWLAAoaGhCA8Ph5OTE5ycnDB06FAMGjSo1vpAxcXFwd3d3RKeAGDw4MFQKBSIj4+3uU5GRgbi4+Oh1+vRp08fNGvWDGFhYTh06FCl+8rOzrYaooGIiIjobrJv4anVamzatAkLFizAqVOn4OTkhM6dO8Pf37826gMApKenlxvlXKVSwdPTE+np6TbXKbuC9MYbb2DJkiXo1q0b1q5di/DwcJw5cwZt2rQpt05ycjI+/PDDKq+iFRcXo7i42PI+JydH7iERERFRAyb7ClSZtm3b4qmnnsJjjz12z+EpNjbWMrp5Ra9z587d07bNZjMAYOrUqZg4cSK6d++OpUuXol27dvjiiy/KLZ+amorIyEg89dRTiImJqXTbCxcuhJubm+Xl5+d3TzUSERFRw1StK1CzZ8/GggUL4OLigtmzZ1e67Pvvv1/tnb/88suIjo6udJnAwEB4e3sjIyPDarrRaERmZia8vb1trufj4wMA6Nixo9X0Dh06lPu24LVr1/DII4+gT58++OSTT6qse+7cuVbtkJOTwxBFRET0AKlWgPr1119hMBgs/66I3NHJvby84OXlVeVyvXv3RlZWFk6cOIHg4GAApQN6ms1mhIaG2lwnICAAvr6+SExMtJp+/vx5DBs2zPI+NTUVjzzyCIKDg7Fq1SrLN/Qqo9FooNFoqlyOiIiIGifZ38Kzl2HDhuH69etYuXKlZRiDkJAQyzAGqampCA8Px9q1a9GzZ08AwLJlyzB//nx8/vnn6NatG9asWYMlS5bgzJkzCAoKQmpqKgYOHAh/f3+sWbMGSqXSsr+KrmzZwm/hERERNTz38/ktuxO5vaxfvx4zZ85EeHg4FAoFRo8ejeXLl1vmGwwGJCYmoqCgwDLtpZdeQlFREWbNmoXMzEx07doVe/bsQVBQEABgz549SE5ORnJyMlq0aGG1vwaSK4mIiMgOqnUF6oknnqj2Brdu3XpfBTVEvAJFRETU8NT6OFB3fuNMp9Phhx9+wPHjxy3zT5w4gR9++AFubm7yKiciIiJqgKp1C2/VqlWWf7/22msYM2YMVq5caekzZDKZMH36dF59ISIiogeC7E7kXl5eOHToENq1a2c1PTExEX369MHNmzdrtMCGgLfwiIiIGp46fZSL0Wi0Obhl2YN5iYiIiBo72d/CmzhxIiZPnowLFy5YhguIj4/HokWLMHHixBovkIiIiKi+kR2glixZAm9vb7z33ntIS0sDUDrq9yuvvIKXX365xgskIiIiqm/uayDNsofoPuj9ftgHioiIqOGp0z5QQGk/qL179+Lf//635fEt165dQ15e3r1sjoiIiKhBkX0L748//kBkZCQuX76M4uJiDBkyBK6urnj33XdRXFyMlStX1kadRERERPWG7CtQL774IkJCQnDr1i04OTlZpj/++OP44YcfarQ4IiIiovpI9hWon376CT///DPUarXV9ICAAKSmptZYYURERET1lewrUGazGSaTqdz0q1evwtXVtUaKIiIiIqrPZAeooUOHYtmyZZb3kiQhLy8P8+fPx/Dhw2uyNiIiIqJ6SfYwBleuXEFkZCSEEEhKSkJISAiSkpLQtGlT/Pjjj9Dr9bVVa73FYQyIiIganvv5/L6ncaCMRiM2bdqEU6dOIS8vDz169EBUVJRVp/IHCQMUERFRw1NnAcpgMKB9+/bYsWMHOnToILvQxooBioiIqOGps4E0HRwcUFRUJGsHRERERI2N7E7kM2bMwLvvvguj0Vgb9RARERHVe7LHgTp27Bh++OEHfP/99+jcuTNcXFys5m/durXGiiMiIiKqj2QHKHd3d4wePbo2aiEiIiJqEGQHqFWrVtVGHUREREQNRrX7QJnNZrz77rvo27cvHn74YcTGxqKwsLA2ayMiIiKql6odoN5++2387W9/g1arRfPmzfHBBx9gxowZtVkbERERUb1U7QC1du1a/Otf/8Lu3bvxzTffYPv27Vi/fj3MZnNt1kdERERU71Q7QF2+fNnqWXeDBw+GJEm4du1arRRGREREVF9VO0AZjUY4OjpaTXNwcIDBYKjxooiIiIjqs2p/C08IgejoaGg0Gsu0oqIiPPfcc1ZjQXEcKCIiImrsqh2gJkyYUG7a008/XaPFEBERETUE1Q5QHP+JiIiIqJTsZ+ERERERPegYoIiIiIhkYoAiIiIikokBioiIiEgmBigiIiIimRigiIiIiGRigCIiIiKSiQGKiIiISCYGKCIiIiKZGKCIiIiIZGKAIiIiIpKJAYqIiIhIJgYoIiIiIpkYoIiIiIhkYoAiIiIikokBioiIiEgmBigiIiIimRigiIiIiGRigCIiIiKSiQGKiIiISCYGKCIiIiKZGKCIiIiIZGKAIiIiIpKJAYqIiIhIJgYoIiIiIpkYoIiIiIhkYoAiIiIikokBioiIiEgmBigiIiIimRigiIiIiGRigCIiIiKSqcEEqMzMTERFRUGn08Hd3R2TJ09GXl5elevFxcVh0KBBcHFxgU6nw4ABA1BYWFhuueLiYnTr1g2SJOHkyZO1cARERETUWDSYABUVFYWzZ89iz5492LFjB3788UdMmTKl0nXi4uIQGRmJoUOH4ujRozh27BhmzpwJhaL8Yb/66qvw9fWtrfKJiIioEZGEEMLeRVQlISEBHTt2xLFjxxASEgIA2LVrF4YPH46rV69WGHx69eqFIUOGYMGCBZVuf+fOnZg9eza+/vprPPTQQ/j111/RrVu3ateXk5MDNzc3ZGdnQ6fTVXs9IiIisp/7+fxuEFeg4uLi4O7ubglPADB48GAoFArEx8fbXCcjIwPx8fHQ6/Xo06cPmjVrhrCwMBw6dMhquevXryMmJgbr1q2Ds7NzteopLi5GTk6O1YuIiIgeHA0iQKWnp0Ov11tNU6lU8PT0RHp6us11Ll68CAB44403EBMTg127dqFHjx4IDw9HUlISAEAIgejoaDz33HNW4awqCxcuhJubm+Xl5+d3j0dGREREDZFdA1RsbCwkSar0de7cuXvattlsBgBMnToVEydORPfu3bF06VK0a9cOX3zxBQDgww8/RG5uLubOnStr23PnzkV2drbldeXKlXuqkYiIiBomlT13/vLLLyM6OrrSZQIDA+Ht7Y2MjAyr6UajEZmZmfD29ra5no+PDwCgY8eOVtM7dOiAy5cvAwD27duHuLg4aDQaq2VCQkIQFRWFNWvW2Ny2RqMptw4RERE9OOwaoLy8vODl5VXlcr1790ZWVhZOnDiB4OBgAKXhx2w2IzQ01OY6AQEB8PX1RWJiotX08+fPY9iwYQCA5cuX4x//+Idl3rVr1xAREYFNmzZVuF0iIiIiuwao6urQoQMiIyMRExODlStXwmAwYObMmRg3bpzlG3ipqakIDw/H2rVr0bNnT0iShFdeeQXz589H165d0a1bN6xZswbnzp3Dli1bAAAtW7a02o9WqwUABAUFoUWLFnV7kERERNRgNIgABQDr16/HzJkzER4eDoVCgdGjR2P58uWW+QaDAYmJiSgoKLBMe+mll1BUVIRZs2YhMzMTXbt2xZ49exAUFGSPQyAiIqJGokGMA1XfcRwoIiKihqfRjwNFREREVJ8wQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUEREREQyMUARERERyaSydwFkm9kskJpViPwSI1zUKjR3d4JCIdm7LCKqZfb+3be1fwC4eqsAF//MBwC0auoCPw/natdV1THdPd9H54i0nCJZbVDdditbLrfYgLwiI7QaFbQaFQSAQoPpnurz0TniWnahpX1auDvi9/RcZOQUw9tNg6HtvaFWK2XVWVfMZoErtwpw6XbtAZ7OkBRShW1Rl3WlZhUit8iAvGIjtI4quGoc7N5ed2owASozMxPPP/88tm/fDoVCgdGjR+ODDz6AVqutdL24uDj8/e9/R3x8PJRKJbp164bdu3fDycnJssx3332Ht956C6dPn4ajoyPCwsLwzTff1PIRVSw5Ixe7z1zHhRt5KDKa4KhSIshLi4hOzdBa72q3uoiodtn7d9/W/t2dHHCrsASJ6bnILjBASIC7kxq9Wnnir71aVllXVcd09/wSoxnFBjM0DgqoVYpqtUF1261suV+v3MLlzAIUlpigVEhQShI0Dko01arRVKuRVV+J0Yxb+SXIKjSgqMSEvBIjCoqNEAJQKCSoFBIWu57Hs/1boXdQk3r1tz05IxcbjlzGkUuZyCosgdFkhhCAzqk0qNzdFnVZl+XndLMAhQYTnNRKtPR0Rnc/j3rzWSgJIYS9i6iOYcOGIS0tDR9//DEMBgMmTpyIhx9+GBs2bKhwnbi4OERGRmLu3LkYMWIEVCoVTp06hZEjR0Kj0QAAvv76a8TExOCdd97BoEGDYDQacebMGYwZM6bateXk5MDNzQ3Z2dnQ6XT3dZzJGblYdTgFmfkl8HFzhLNahYISI9Kyi+DposbEvgH14sQhoppl7999W/u/llWAny9mIrfIADdHBzR1VUOChKwCA8xCoKufO14a3KbSYFPZMQ1qr8e+cxmW+UUGM078kYmsQgM8nB3Qo6UHHB2UlbZBddutbLnLNwuQkVsEo0kAkkBGTjEEAJ2jAzxc1GjXTItCg7mC+kz45fIt3CowwN3JAa31rvj9WjbSc4qgVEhwdlDiRl4JBEr7x6iUgJODEkVGMxyUSrT31sLL1bFe/G1PzsjFsr1JOHUlCwpJgqODAjdyi1BoMEOpkOCtc0KXFjpLW9RVfZafU2YBbuQWwWA0Q+OgRLHBDJVSgt7VES2bONdYPffz+d0grkAlJCRg165dOHbsGEJCQgAAH374IYYPH44lS5bA19fX5nqzZs3CCy+8gNjYWMu0du3aWf5tNBrx4osvYvHixZg8ebJleseOHWvpSCpnNgvsPnMdmfklaKPXQpJKL1O6OjpAq1EhKSMP35+9jsCm2npzCZOI7p+9f/dt7V8IgbTsIpQYTBBmAYUEOKqUkCQJzXQK3Mwvwfnrudh9Jh2BA8vXVdUxnb+ei9U/p8BFrULbZqV3EhLSbsFkFmjp4YRbBQak3CxAiL8H2ui1Ntuguu0W4OmC3Weu42ZeMYxmM0xmAU8XB6RlF0OlkABJgiQBRQYT0nOKEdzSHUkZeeXqO5eWC6OptL7M/BKcvnoL+SUmaFQKQAA380sAACoJEABM5tKXh5MDbuSV4Oy1HEzq4wGVSlWnP19bP+9dv6Xj/PVcqFUKeDqXtoWABA9nBxSUmJBZUIy07CKE+Hsg+UZ+ndRX9vO8mVcCo9EMo0mgiVYDSZKg1Qhk5pfAaDbjZl5xvfgsbBCdyOPi4uDu7m4JTwAwePBgKBQKxMfH21wnIyMD8fHx0Ov16NOnD5o1a4awsDAcOnTIsswvv/yC1NRUKBQKdO/eHT4+Phg2bBjOnDlTaT3FxcXIycmxetWE1KxCXLiRBx83R8sfgjKSJMHHzRHJGXlIzSqskf0RUf1g7999W/vPLTIiI6cYZgE4a1QoNJhRYjRbanJ1VMFkFjh9NdtmXVUdk6ujCil/5kPnqIIkScgtMuJWQQm0jg5QKBTQOqqQmV+C3CJjhW1Q3Xb75cotXLiRB1dHFW4VGKB1dIDBJFBoMEHtoIRapUCRwQyNUoHM/BLkFZts1pdZUAKtowoKhQJqByVu5RtgMgloHJQQEmASgEIqvXVX9rluMJthNANKCSgxmXHxz4IK66yrv+2pWYX4LTUbJrOA651toVJAun0702wGMnKKkVdsqrP6yn6eOkcVbhUaoL3d9kBpO2lv//xcHVX14rOwQQSo9PR06PV6q2kqlQqenp5IT0+3uc7FixcBAG+88QZiYmKwa9cu9OjRA+Hh4UhKSiq3zP/8z/9gx44d8PDwwMCBA5GZmVlhPQsXLoSbm5vl5efnVxOHifwSI4qMJjirbV8YdFIrUWw0Ib/EWCP7I6L6wd6/+7b2X2Iyw2AyAxBwUEowCwHTHT0+HJSlHx8FBqPNuqo6JqVCgsFktlxBKDGZYTSZ4aCULNs3mc0oMZWGNlttUN12u5lfgiKjCUqFwrIPkxAwCwGlVNoHyiwEJAVgvL1Pm/WZzZbjVkjW2zBX0BlGCMAsBCABEECujbaq67/t+SVG5BtK93V3WwC4/V8Bw+22qKv6yn6eCoVk1dZlHJQKGM2lP5v68Flo1wAVGxsLSZIqfZ07d+6etm02l/7STZ06FRMnTkT37t2xdOlStGvXDl988YXVMn//+98xevRoBAcHY9WqVZAkCZs3b65w23PnzkV2drbldeXKlXuq8W4uahUcVUoUVHBSFJaYoFEp4VLBHwsiapjs/btva/9qpeL2B5gEg0lAcTtolDHcDjbODiqbdVV1TCazgINSAfPt5KFWKqBSKmAwCcv2lQoF1Lc/RG21QXXbrYmLGo4qJUxms2UfSkmyhCCTKD0+YQZUt/dpsz6FwnLcZYGjbBsV3UmSpNKwBQFAAlxttFVd/213Uavg4lC6r7vbAsDt/0pwuN0WdVVf2c/TbBZWbV3GYDJDpSj92dSHz0K7BqiXX34ZCQkJlb4CAwPh7e2NjIwMq3WNRiMyMzPh7e1tc9s+Pj4Ayvdn6tChAy5fvlzhMhqNBoGBgZZlbNFoNNDpdFavmtDc3QlBXlqkZRfh7r79Zf0RWuu1lq8VE1HjYO/ffVv7d3VUQa/TQCEBBcVGON3+1llZTblFRigVErq0cLNZV1XHlFtkREBTF+QUGSGEgKujCh7OauQVGWA2m5FXZISnixqujqoK26C67dbDzwNBXlrkFhnh4eyAvCIDHJQSnByUKDGUfrPO0UGBYlNph2mtRmmzPk9nNfKKjDCbzSgxmODh4gClUkKxwQRJlN6mM4vSvjxlV6QcFAqoFKW399RKBQKbOldYZ139bW/u7oTOzd2gVEjIvbMtjGYIIVBsMEGhAPQ6DbQaZZ3VV/bzzCkywsPJAXm32x4obae82z+/3CJjvfgstGuA8vLyQvv27St9qdVq9O7dG1lZWThx4oRl3X379sFsNiM0NNTmtgMCAuDr64vExESr6efPn4e/vz8AIDg4GBqNxmoZg8GAlJQUyzJ1SaGQENGpGTxd1EjKyENukQFGsxm5RQYkZeTB00WNoQ81YwdyokbG3r/7tvZvEgI+bo5QOyghKUpvcRUZTSg0mHA9pxglRjPaNnNFRCdvm3VVdUxNtBpE9wlAE23p/LxiIwKaOkOpkHD5ViGUCgkBTZyRV2yssA2q224qlQIRnZqhiVYDlUIBpUJCZr4BGgcJRrOA4fbX9x0dlPDWaZB8I7/K+lQqJTq3cIebowOKjWYYhUATFzUAwChKb90pFaWh6lahAU5qFR7y1eHizUK7/21XKCREdvZG22auKDGakZFbAmeNEhIEbhUYYDQLeLqoS/s+3civs/rKfp5NtGqoVKU/p5t5xcgtMuBmXgmUt68CNtFq6sVnYYMaxuD69etYuXKlZRiDkJAQyzAGqampCA8Px9q1a9GzZ08AwLJlyzB//nx8/vnn6NatG9asWYMlS5bgzJkzCAoKAgC89NJL2LJlC7744gv4+/tj8eLF2L59O86dOwcPD49q1VaTwxgA1mOaFBtLL5221msx9KH6MfYFEdUOe//u29q/h7MDMgtujwNVaAAAuDmp0TvQE+ND5Y0DZeuY7p5fbDSj2GiGRqW4/aq6DarbbneOL3QlswAFFYwDVWV9t8eB0qgUKL5zHCiDCXnF5ceB0uscMbmf9ThQ9eFv+53jQGUXlliC5J3jQNmjPlvjdTmrlfDzdEaPlh41Ws/9fH43mACVmZmJmTNnWg2kuXz5cstAmikpKWjVqhX279+PgQMHWtZbtGgRVqxYgczMTHTt2hX//Oc/0a9fP8t8g8GAuXPnYt26dSgsLERoaCiWLVuGhx56qNq11XSAAurfaLVEVDfs/bvPkcg5EvmDNBL5AxGg6rPaCFBERERUu+7n87tBDGNAREREVJ/w+/BERETVUN9uv5F9MUARETUw/CCvXbba9+Kfedh1Jh2/pWajoMQIZ7UKnZu7IbKTt12/3FPZucDzpHYxQBERNSB3fiusyGiCo0qJIC9tvXlCfUNnq33dnRxw4c98pGUV3h5ksnSgyUs38nEuPbfSByrXda1l5wIAnie1jJ3IawA7kRNRXSh7Un1mfgl83BzhrFahoMSItOwieLqoa+wJ9Q8qW+2bX2zAvsQb+DOvGJ5ODmjiqoGDsnSU7NxCA0pMAkM6NsP/PNqxTq/uVHYuKG/XYTILnidVYCdyIqJGruxJ9Zn5JWij18LV0QFKhQRXRwe00WuRmV+C789etzx6hOSpqH2FAPKLjDCbBQRKRxNXSFLpI2K0pSO1x1+8iau3CqrcR23X6urogNZeLjifnovz13PR2suF50ktYoAiImoAyp5U7+PmaHlCfRlJkkpHja4HT6hvqCpq31sFhtsPLFai0GBGifH/ns8mSRLcnB2QVWiwjAFlz1oBIK/YVPp8P7NAXrHJah7Pk5rFAEVE1ACUPaneuYIHqDqplfXiCfUNVYXtezufKG8/zsZUrtdL3XfKruxcKDGZUdpHq+zf1nie1BwGKCKiBqDsSfUFFXzwFZaY6sUT6huqitrXw8kBjg5KFJaYIAFQ3nHFRwiB7AID3JzUaNXUxe61AqW3GMtCXem/rfE8qTkMUEREDUDZk+rTsotw93d/hBBIyy6qF0+ob6gqal+dkwOauznBaBaAVHptxywEio0m3MwvgVkI9A70hJ+Hs91rBQCtRgmlJEGpkKDVKK3m8TypWYygREQNQNmT6q9lFyIpo7T/i5O69MpI2ber6sMT6huqytpX5+wAvasjACCv2AgUl66jVEjo6ueO8aEt67TdqzoX2nqXfsMu+UY+z5NaxGEMagCHMSCiunLn2D/FxtLbMa312hp9Qv2DrKL2beftioRrOaUDaRqMcHZQoUsLN0TYcSDNys4FADxPqoEPE7YzBigiqkscYbp2VdS+9bHdORL5/bmfz2/ewiMiamAUCgl+nnXX5+ZBU1H71sd2r6ym+lhvY8JO5EREREQyMUARERERycQARURERCQTAxQRERGRTAxQRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMnEk8hpQ9jScnJwcO1dCRERE1VX2uX0vT7VjgKoBubm5AAA/Pz87V0JERERy5ebmws3NTdY6fJhwDTCbzbh27RpcXV0hSfZ/UGNOTg78/Pxw5cqVB/bhxmwDtgHANgDYBgDbAGAbALbbQAiB3Nxc+Pr6QqGQ16uJV6BqgEKhQIsWLexdRjk6ne6B/UUpwzZgGwBsA4BtALANALYBUL4N5F55KsNO5EREREQyMUARERERycQA1QhpNBrMnz8fGo3G3qXYDduAbQCwDQC2AcA2ANgGQM23ATuRExEREcnEK1BEREREMjFAEREREcnEAEVEREQkEwMUERERkUwMUI3QihUrEBAQAEdHR4SGhuLo0aP2LqnOvPHGG5AkyerVvn17e5dVq3788UeMGDECvr6+kCQJ33zzjdV8IQTmzZsHHx8fODk5YfDgwUhKSrJPsbWkqjaIjo4ud15ERkbap9hasnDhQjz88MNwdXWFXq/HqFGjkJiYaLVMUVERZsyYgSZNmkCr1WL06NG4fv26nSquedVpg4EDB5Y7F5577jk7VVzzPvroI3Tp0sUyWGTv3r2xc+dOy/zGfg4AVbdBTZ0DDFCNzKZNmzB79mzMnz8fv/zyC7p27YqIiAhkZGTYu7Q689BDDyEtLc3yOnTokL1LqlX5+fno2rUrVqxYYXP+P//5TyxfvhwrV65EfHw8XFxcEBERgaKiojqutPZU1QYAEBkZaXVe/Pvf/67DCmvfwYMHMWPGDBw5cgR79uyBwWDA0KFDkZ+fb1lm1qxZ2L59OzZv3oyDBw/i2rVreOKJJ+xYdc2qThsAQExMjNW58M9//tNOFde8Fi1aYNGiRThx4gSOHz+OQYMGYeTIkTh79iyAxn8OAFW3AVBD54CgRqVnz55ixowZlvcmk0n4+vqKhQsX2rGqujN//nzRtWtXe5dhNwDEtm3bLO/NZrPw9vYWixcvtkzLysoSGo1G/Pvf/7ZDhbXv7jYQQogJEyaIkSNH2qUee8nIyBAAxMGDB4UQpT93BwcHsXnzZssyCQkJAoCIi4uzV5m16u42EEKIsLAw8eKLL9qvKDvw8PAQn3322QN5DpQpawMhau4c4BWoRqSkpAQnTpzA4MGDLdMUCgUGDx6MuLg4O1ZWt5KSkuDr64vAwEBERUXh8uXL9i7Jbi5duoT09HSrc8LNzQ2hoaEP1DkBAAcOHIBer0e7du0wbdo03Lx5094l1ars7GwAgKenJwDgxIkTMBgMVudC+/bt0bJly0Z7LtzdBmXWr1+Ppk2bolOnTpg7dy4KCgrsUV6tM5lM2LhxI/Lz89G7d+8H8hy4uw3K1MQ5wIcJNyJ//vknTCYTmjVrZjW9WbNmOHfunJ2qqluhoaFYvXo12rVrh7S0NLz55pvo378/zpw5A1dXV3uXV+fS09MBwOY5UTbvQRAZGYknnngCrVq1woULF/C3v/0Nw4YNQ1xcHJRKpb3Lq3FmsxkvvfQS+vbti06dOgEoPRfUajXc3d2tlm2s54KtNgCAv/71r/D394evry9Onz6N1157DYmJidi6dasdq61Zv/32G3r37o2ioiJotVps27YNHTt2xMmTJx+Yc6CiNgBq7hxggKJGZdiwYZZ/d+nSBaGhofD398dXX32FyZMn27Eysqdx48ZZ/t25c2d06dIFQUFBOHDgAMLDw+1YWe2YMWMGzpw50+j7/1WmojaYMmWK5d+dO3eGj48PwsPDceHCBQQFBdV1mbWiXbt2OHnyJLKzs7FlyxZMmDABBw8etHdZdaqiNujYsWONnQO8hdeING3aFEqlstw3Kq5fvw5vb287VWVf7u7uaNu2LZKTk+1dil2U/dx5TlgLDAxE06ZNG+V5MXPmTOzYsQP79+9HixYtLNO9vb1RUlKCrKwsq+Ub47lQURvYEhoaCgCN6lxQq9Vo3bo1goODsXDhQnTt2hUffPDBA3UOVNQGttzrOcAA1Yio1WoEBwfjhx9+sEwzm8344YcfrO79Pkjy8vJw4cIF+Pj42LsUu2jVqhW8vb2tzomcnBzEx8c/sOcEAFy9ehU3b95sVOeFEAIzZ87Etm3bsG/fPrRq1cpqfnBwMBwcHKzOhcTERFy+fLnRnAtVtYEtJ0+eBIBGdS7czWw2o7i4+IE4BypS1ga23PM5cN/d0Kle2bhxo9BoNGL16tXi999/F1OmTBHu7u4iPT3d3qXViZdfflkcOHBAXLp0SRw+fFgMHjxYNG3aVGRkZNi7tFqTm5srfv31V/Hrr78KAOL9998Xv/76q/jjjz+EEEIsWrRIuLu7i2+//VacPn1ajBw5UrRq1UoUFhbaufKaU1kb5Obmijlz5oi4uDhx6dIlsXfvXtGjRw/Rpk0bUVRUZO/Sa8y0adOEm5ubOHDggEhLS7O8CgoKLMs899xzomXLlmLfvn3i+PHjonfv3qJ37952rLpmVdUGycnJ4q233hLHjx8Xly5dEt9++60IDAwUAwYMsHPlNSc2NlYcPHhQXLp0SZw+fVrExsYKSZLE999/L4Ro/OeAEJW3QU2eAwxQjdCHH34oWrZsKdRqtejZs6c4cuSIvUuqM2PHjhU+Pj5CrVaL5s2bi7Fjx4rk5GR7l1Wr9u/fLwCUe02YMEEIUTqUweuvvy6aNWsmNBqNCA8PF4mJifYtuoZV1gYFBQVi6NChwsvLSzg4OAh/f38RExPT6P6nwtbxAxCrVq2yLFNYWCimT58uPDw8hLOzs3j88cdFWlqa/YquYVW1weXLl8WAAQOEp6en0Gg0onXr1uKVV14R2dnZ9i28Bk2aNEn4+/sLtVotvLy8RHh4uCU8CdH4zwEhKm+DmjwHJCGEkHfNioiIiOjBxj5QRERERDIxQBERERHJxABFREREJBMDFBEREZFMDFBEREREMjFAEREREcnEAEVEREQkEwMUEdEDLCAgAMuWLbN3GUQNDgMU0QNKkqRKX2+88Uad1pOcnIxJkyahZcuW0Gg0aN68OcLDw7F+/XoYjcY6rcWWgIAAm+20aNEie5d2X44dO2b1dHoiqh6VvQsgIvtIS0uz/HvTpk2YN28eEhMTLdO0Wq3l30IImEwmqFS18yfj6NGjGDx4MB566CGsWLEC7du3BwAcP34cK1asQKdOndC1a9da2bccb731FmJiYqymubq61uo+DQYDHBwcam37Xl5etbZtosaMV6CIHlDe3t6Wl5ubGyRJsrw/d+4cXF1dsXPnTgQHB0Oj0eDQoUOIjo7GqFGjrLbz0ksvYeDAgZb3ZrMZCxcuRKtWreDk5ISuXbtiy5YtFdYhhEB0dDTatm2Lw4cPY8SIEWjTpg3atGmD8ePH49ChQ+jSpYtl+StXrmDMmDFwd3eHp6cnRo4ciZSUFMv8shqXLFkCHx8fNGnSBDNmzIDBYLAsU1xcjDlz5qB58+ZwcXFBaGgoDhw4UGWbubq6WrWbt7c3XFxcAJSGK19fX9y8edOy/KOPPopHHnkEZrMZQOlVv48++gjDhg2Dk5MTAgMDrdomJSUFkiRh06ZNCAsLg6OjI9avXw8A+Oyzz9ChQwc4Ojqiffv2+Ne//mVZr6SkBDNnzoSPjw8cHR3h7++PhQsXWtr3jTfesFzZ8/X1xQsvvGBZ9+5beJcvX8bIkSOh1Wqh0+kwZswYXL9+3TL/jTfeQLdu3bBu3ToEBATAzc0N48aNQ25ubpXtR9SYMEARUYViY2OxaNEiJCQkWIWYyixcuBBr167FypUrcfbsWcyaNQtPP/00Dh48aHP5kydPIiEhAXPmzIFCYftPkiRJAEqvxkRERMDV1RU//fQTDh8+DK1Wi8jISJSUlFiW379/Py5cuID9+/djzZo1WL16NVavXm2ZP3PmTMTFxWHjxo04ffo0nnrqKURGRiIpKamaLVPe3//+dwQEBODZZ58FAKxYsQI///wz1qxZY3Vcr7/+OkaPHo1Tp04hKioK48aNQ0JCgtW2YmNj8eKLLyIhIQERERFYv3495s2bh7fffhsJCQl455138Prrr2PNmjUAgOXLl+M///kPvvrqKyQmJmL9+vUICAgAAHz99ddYunQpPv74YyQlJeGbb75B586dbR6D2WzGyJEjkZmZiYMHD2LPnj24ePEixo4da7XchQsX8M0332DHjh3YsWMHDh482OBvZRLJVoMPQCaiBmrVqlXCzc3N8n7//v0CgPjmm2+slpswYYIYOXKk1bQXX3xRhIWFCSGEKCoqEs7OzuLnn3+2Wmby5Mli/PjxNve9ceNGAUD88ssvlmnXr18XLi4ulteKFSuEEEKsW7dOtGvXTpjNZsuyxcXFwsnJSezevdtSo7+/vzAajZZlnnrqKTF27FghhBB//PGHUCqVIjU11aqO8PBwMXfu3IqayPJ09zvrcnFxET/++KNlmQsXLghXV1fx2muvCScnJ7F+/XqrbQAQzz33nNW00NBQMW3aNCGEEJcuXRIAxLJly6yWCQoKEhs2bLCatmDBAtG7d28hhBDPP/+8GDRokFW7lHnvvfdE27ZtRUlJSYXHtXTpUiGEEN9//71QKpXi8uXLlvlnz54VAMTRo0eFEELMnz9fODs7i5ycHMsyr7zyiggNDbW5faLGin2giKhCISEhspZPTk5GQUEBhgwZYjW9pKQE3bt3r/Z2mjRpgpMnTwIABg4caLm6dOrUKSQnJ5frd1RUVIQLFy5Y3j/00ENQKpWW9z4+Pvjtt98AAL/99htMJhPatm1rtY3i4mI0adKk0rpeeeUVREdHW01r3ry55d+BgYFYsmQJpk6dirFjx+Kvf/1ruW307t273PuyYy1zZ7vn5+fjwoULmDx5slX/K6PRCDc3NwClty2HDBmCdu3aITIyEo899hiGDh0KAHjqqaewbNkyBAYGIjIyEsOHD8eIESNs9mdLSEiAn58f/Pz8LNM6duwId3d3JCQk4OGHHwZQetvvzp+Bj48PMjIybLYZUWPFAEVEFSrr31NGoVBACGE17c6+RXl5eQCA7777zipYAIBGo7G5jzZt2gAAEhMTLSFLqVSidevWAGD1QZ+Xl4fg4GBLv6A73dkZ+u5O15IkWfoh5eXlQalU4sSJE1YhC7DuOG9L06ZNLXVV5Mcff4RSqURKSgqMRuM9dby/s93L2vTTTz9FaGio1XJl9ffo0QOXLl3Czp07sXfvXowZMwaDBw/Gli1b4Ofnh8TEROzduxd79uzB9OnTsXjxYhw8ePCeO6dX1r5EDwr2gSKiavPy8rL69h4Aq6snHTt2hEajweXLl9G6dWur151XNe7UvXt3tG/fHkuWLKnyQ7hHjx5ISkqCXq8vt/2yqzFV6d69O0wmEzIyMsptw9vbu1rbqMimTZuwdetWHDhwAJcvX8aCBQvKLXPkyJFy7zt06FDhNps1awZfX19cvHixXL2tWrWyLKfT6TB27Fh8+umn2LRpE77++mtkZmYCAJycnDBixAgsX74cBw4cQFxcnOWK3J06dOiAK1eu4MqVK5Zpv//+O7KystCxY0fZ7UHUmPEKFBFV26BBg7B48WKsXbsWvXv3xpdffokzZ85Yrhy5urpizpw5mDVrFsxmM/r164fs7GwcPnwYOp0OEyZMKLdNSZKwatUqDBkyBH379sXcuXPRoUMHGAwG/Pjjj7hx44blSktUVBQWL16MkSNH4q233kKLFi3wxx9/YOvWrXj11VfRokWLKo+hbdu2iIqKwjPPPIP33nsP3bt3x40bN/DDDz+gS5cuePTRRytcNzc3F+np6VbTnJ2dodPpcPXqVUybNg3vvvsu+vXrh1WrVuGxxx7DsGHD0KtXL8vymzdvRkhICPr164f169fj6NGj+Pzzzyut+c0338QLL7wANzc3REZGori4GMePH8etW7cwe/ZsvP/++/Dx8UH37t2hUCiwefNmeHt7w93dHatXr4bJZEJoaCicnZ3x5ZdfwsnJCf7+/uX2M3jwYHTu3BlRUVFYtmwZjEYjpk+fjrCwMNm3c4kaPXt3wiIi+6uoE/mtW7fKLTtv3jzRrFkz4ebmJmbNmiVmzpxp6UQuhBBms1ksW7ZMtGvXTjg4OAgvLy8REREhDh48WGkNiYmJYsKECaJFixZCpVIJNzc3MWDAAPHxxx8Lg8FgWS4tLU0888wzomnTpkKj0YjAwEARExMjsrOzhRBVd3QXQoiSkhIxb948ERAQIBwcHISPj494/PHHxenTpyusz9/fXwAo95o6daowm80iPDxcREREWHXkfv7550VQUJDIzc0VQpR2Il+xYoUYMmSI0Gg0IiAgQGzatMmyfFkn8l9//bXc/tevXy+6desm1Gq18PDwEAMGDBBbt24VQgjxySefiG7dugkXFxeh0+lEeHi4pVP+tm3bRGhoqNDpdMLFxUX06tVL7N271+q4yjqRC1Hayf4vf/mLcHFxEa6uruKpp54S6enplvnz588XXbt2tapt6dKlwt/fv8K2I2qMJCHu6tBARES1QpIkbNu2rdxYWkTU8LAPFBEREZFMDFBEREREMrETORFRHWGPCaLGg1egiIiIiGRigCIiIiKSiQGKiIiISCYGKCIiIiKZGKCIiIiIZGKAIiIiIpKJAYqIiIhIJgYoIiIiIpkYoIiIiIhk+v8FX4c9h/ud9gAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.scatter(merged[\"TrueValue\"], merged[\"SCORE\"], alpha=0.5)\n",
    "plt.xlabel(\"True Gene Expression\")\n",
    "plt.ylabel(\"Predicted PRS\")\n",
    "plt.title(\"PRS vs True Gene Expression\")\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "myenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
